<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>singletCode PyPI package &mdash; singletCode 1.2 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/sphinx_rtd_size.css" type="text/css" />
      <link rel="stylesheet" href="../_static/colour.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script type="text/javascript">document.addEventListener('DOMContentLoaded', function() {document.querySelectorAll('a.reference.external').forEach(function(a) {a.setAttribute('target', '_blank'); a.setAttribute('rel', 'noopener noreferrer');});});</script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex/" />
    <link rel="search" title="Search" href="../search/" />
    <link rel="next" title="command line interface" href="../infoCLI/" />
    <link rel="prev" title="dataset information" href="../datasetsInfo/" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../" class="icon icon-home">
            singletCode
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search/" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../doubletInformation/">doublets in sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../singletCode/">what is singletCode?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../classifier/">classifier</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usageInformation/">implementing singletCode on your data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../doubletDetectionMethods/">benchmarking doublet detection methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../datasetsInfo/">dataset information</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">singletCode PyPI package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#availablility">Availablility</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#detailed-information-about-the-parameters-for-the-functions-in-the-package">Detailed information about the parameters for the functions in the package</a></li>
<li class="toctree-l2"><a class="reference internal" href="#get-singlets">get_singlets</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#source.singletCode.singletCode.get_singlets"><code class="docutils literal notranslate"><span class="pre">get_singlets()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#check-sample-sheet">check_sample_sheet</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#source.singletCode.singletCode.check_sample_sheet"><code class="docutils literal notranslate"><span class="pre">check_sample_sheet()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../infoCLI/">command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../watermelonDatasetVignette/">Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../singletCodePackageVignette/">Vignette to use singletCode package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Contact/">questions? contact us!</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.cell.com/cell-genomics/fulltext/S2666-979X(24)00176-9">link to paper</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.goyallab.org/">Goyal Lab website</a></li>
<li class="toctree-l1"><a class="reference internal" href="../repositories/">repositories</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../">singletCode</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">singletCode PyPI package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/packageDocumentation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="singletcode-pypi-package">
<h1>singletCode PyPI package<a class="headerlink" href="#singletcode-pypi-package" title="Permalink to this heading">¶</a></h1>
<section id="availablility">
<h2>Availablility<a class="headerlink" href="#availablility" title="Permalink to this heading">¶</a></h2>
<p>The package is available on PyPI as <a class="reference external" href="https://pypi.org/project/singletCode/">singletCode</a> .</p>
</section>
<section id="how-do-i-use-it">
<span id="package-info"></span><h2>How do I use it?<a class="headerlink" href="#how-do-i-use-it" title="Permalink to this heading">¶</a></h2>
<blockquote>
<div><ol class="arabic">
<li><p>Installation</p>
<blockquote>
<div><p>It can be installed from PyPI using the following in the terminal:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pip3</span> <span class="n">install</span> <span class="n">singletCode</span>
<span class="kn">from</span> <span class="nn">singletCode</span> <span class="kn">import</span> <span class="n">check_sample_sheet</span><span class="p">,</span> <span class="n">get_singlets</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Preparing the input sample sheet.</p>
<blockquote>
<div><p>The input sample sheet is a .csv file that contains the information about cell ID (added while sequencing), lineage barcode, and sample name. Each row should be repeated n times where n is the number of UMIs associated with that barcode and cell ID combination. For creating the input, you can read the .csv file in as a pandas dataframe.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;path/to/csv/file.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
<p>You can check if the format and the column names are valid for running singletCode using check_sample_sheet() function.</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">check_sample_sheet</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<p>It will either return an error with information about how to modify the sample sheet to make it a valid input or will print the message, “The sample sheet provided can be used as input to get_singlets to get a list of singlets identified”, in which case you can move to the next step.</p>
</div></blockquote>
</li>
<li><p>Running get_singlets() to get an assignment of singlet status for each cell ID and barcode combination.</p>
<blockquote>
<div><p>This is the step where singlet identification is done using the singletCode framework. You can read more about the parameters for get_singlets() below. In this example, default values are used.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cellLabelList</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="n">get_singlets</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;Sample1&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>cellLabelList</strong> is a pandas dataframe that contains 5 rows: cellID, barcode, sample, nUMI and label. Label is whether the particular cell ID and barcode combination has been called a singlet or not.</p>
<p>To get a dataframe containing just the singlets, you can run this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">singletList</span> <span class="o">=</span> <span class="n">cellLabelList</span><span class="p">[</span><span class="n">cellLabelList</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Singlet&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p><strong>stats</strong> contains the statistics for each sample present in your dataset: total cells, total number of singlets, number of singlets recovered from different categories of singlets (such as single-barcode singlets, muli-barcode singlets, dominant-UMI singlets), number of cells removed due to low barcode UMI counts for the barcode and number of indeterminate cells since singletCode can identify only truly singlet cells but not be certain if the other cells are truly not singlets.</p>
<p>You can save the singlets list in a .txt file and the list of cell IDs in each category of singlets by setting <strong>save_all_singlet_categories</strong> parameter to TRUE.</p>
</div></blockquote>
</li>
</ol>
</div></blockquote>
</section>
<section id="detailed-information-about-the-parameters-for-the-functions-in-the-package">
<h2>Detailed information about the parameters for the functions in the package<a class="headerlink" href="#detailed-information-about-the-parameters-for-the-functions-in-the-package" title="Permalink to this heading">¶</a></h2>
</section>
<section id="get-singlets">
<h2>get_singlets<a class="headerlink" href="#get-singlets" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="source.singletCode.singletCode.get_singlets">
<span class="sig-name descname"><span class="pre">get_singlets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sheet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_all_singlet_categories</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_plot_umi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_cutoff_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ratio'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_cutoff_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7.5e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_cutoff_percentile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_umi_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_diff_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_dominant_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/source/singletCode/singletCode/get_singlets/#get_singlets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#source.singletCode.singletCode.get_singlets" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that inputs the sample sheet and other parameters and runs it through count_doublets to get a list of
singlets in the sample. If a row is repeated, it is assumed to reflect the UMI associated with a barcode in
this cell (identified by the cellID).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sheet</strong> (<em>DataFrame</em>) – A dataframe that contains 3 columns: cellID, barcode, sample.</p></li>
<li><p><strong>dataset_name</strong> (<em>str</em>) – The name of the dataset being analysed. It will be in the name of all saved files
and be a column in the singlet_stats sheet returned.</p></li>
<li><p><strong>output_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to store any output files, including plots to show UMI distribution
and what the umi_cutoff used is, csv files containing singlets of different categories. If None,
then the list of singlets will be returned but it won’t contain information about what category of
singlet each cell is. Defaults to None.</p></li>
<li><p><strong>save_all_singlet_categories</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, then singlets of each category are saved separately
in csv files along with all singlets and all non-singlets. Defaults to False.</p></li>
<li><p><strong>save_plot_umi</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, then plots showing UMI distribution indicating the UMI cutoff used
will be saved for each sample. Defaults to False.</p></li>
<li><p><strong>umi_cutoff_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Specify if quality control for barcodes using UMI counts should be based
on “ratio” or “percentile”. Defaults to ‘ratio’.</p></li>
<li><p><strong>umi_cutoff_ratio</strong> (<em>float</em><em>, </em><em>optional</em>) – The ratio used to determine the umi_cutoff if umi_cutoff_method is
“ratio”. Defaults to 3/4e5.</p></li>
<li><p><strong>umi_cutoff_percentile</strong> (<em>float</em><em>, </em><em>optional</em>) – If umi_cutoff_method is “percentile”, then the umi_cutoff will be
the minimum UMI count required to be in the top umi_cutoff_percentile’th percentile. There is no default and if umi_cutoff_method is set to “percentile”, then manually set this parameter.</p></li>
<li><p><strong>min_umi_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – This is the absolute minimum number of UMIs that need to be associated with
a barcode for it to be considered a barcode. However, the actual umi_cutoff used will be the greater
of min_umi_cutoff and the cutoff calculated using umi_cutoff_method. Defaults to 2.</p></li>
<li><p><strong>umi_dominant_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum UMI count to be associated with a barcode for it to be
considered to be a potential dominant barcode in a cell. Defaults to 10.</p></li>
<li><p><strong>umi_diff_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – This is the minimum difference between UMI counts associated with a
potential dominant barcode present within a cell and the median UMI count of all barcodes associated
with the cell. If a cell has only one dominant barcode, it will be counted. Defaults to 50.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>A 2-tuple containing:</dt><dd><ul class="simple">
<li><p>pandas.DataFrame: A dataframe which contains all unique cell ID and barcode combinations in the data along with singlet assignment to each cell ID.</p></li>
<li><p>pandas.DataFrame: A dataframe which contains the statistics for total singlets, different categories of singlets, and cells removed due to low UMI counts.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="check-sample-sheet">
<h2>check_sample_sheet<a class="headerlink" href="#check-sample-sheet" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="source.singletCode.singletCode.check_sample_sheet">
<span class="sig-name descname"><span class="pre">check_sample_sheet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_frame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/source/singletCode/singletCode/general_utility_functions/#check_sample_sheet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#source.singletCode.singletCode.check_sample_sheet" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to check if the dataframe can be used as input to get_singlets function. It checks if the three columns - cellID, barcode, sample needed are present and in same order. If this dataframe can be used for get_singlets function, then a statement will be printed to confirm that</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_sheet</strong> – A dataframe that contains your sample sheet.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../datasetsInfo/" class="btn btn-neutral float-left" title="dataset information" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../infoCLI/" class="btn btn-neutral float-right" title="command line interface" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Goyal Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>