<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Datasets used &mdash; SingletCode 1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/sphinx_rtd_size.css" />
      <link rel="stylesheet" type="text/css" href="../_static/_static/custom.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="About" href="about.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            SingletCode
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Datasets used</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SingletCode</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Datasets used</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/source/data.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="datasets-used">
<h1>Datasets used<a class="headerlink" href="#datasets-used" title="Permalink to this heading">¶</a></h1>
<p>SingletCode was tested on datasets with many different barcoding techniques, cell types and sequencing methods. Comprehensive details about the datasets used are in this page.</p>
<table class="longtable docutils align-default" id="id3">
<caption><span class="caption-text">Datasets</span><a class="headerlink" href="#id3" title="Permalink to this table">¶</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Tech</p></th>
<th class="head"><p>Barcode length (bp)</p></th>
<th class="head"><p>total unique barcodes</p></th>
<th class="head"><p>Separately sequenced barcode?</p></th>
<th class="head"><p>Separately amplified barcode?</p></th>
<th class="head"><p>Filtering metrics - cells/ RNAseq/ATAC-seq</p></th>
<th class="head"><p>Filtering metrics - barcode reads, barcodes</p></th>
<th class="head"><p>Are reads with identical cellID and UMI grouped?</p></th>
<th class="head"><p>Are similar barcodes merged? How?</p></th>
<th class="head"><p>Doublet detection method used?</p></th>
<th class="head"><p>Twins?</p></th>
<th class="head"><p>ATAC data available?</p></th>
<th class="head"><p>compatible sequencing methods?</p></th>
<th class="head"><p>How did we get this data</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>FateMap</p></td>
<td><p>100</p></td>
<td><p>48.9-63.3 <a href="#id1"><span class="problematic" id="id2">*</span></a>10e6</p></td>
<td><p>This can be done using gDNA if scRNAseq is not being done (but not in the datasets we use)</p></td>
<td><p>Yes, primers are used to identify barcode region and cell ID region to amplify separately</p></td>
<td><p>Cells with less than 200 genes, cell-type dependent cutoff for mitochondrial gene content</p></td>
<td><p>searching for sequences complementary to the side reaction library preparation primers, filtering out reads that lack the GFP barcode sequence, have too many repeated nucleotides, or do not meet a phred score cut-off</p></td>
<td><p>Yes and the minimum number of reads is used as a cutoff</p></td>
<td><p>Starcode based on Levenshtein distance (cutoff based on manual inspection)</p></td>
<td><p>stringent gene count cutoff per cell, any cell having more than 1 barcode is removed</p></td>
<td><p>Yes</p></td>
<td><p>No</p></td>
<td><p>10X</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>TREX</p></td>
<td><p>30</p></td>
<td><p>0.94 × 10e6</p></td>
<td><p>No</p></td>
<td><p>Yes, it was amplified in the cDNA and the amplicon library was sequenced</p></td>
<td><p>only cells expressing 500–10,000 genes</p></td>
<td><p>reads:
it aligns to the region that is supposed to be the barcode(clone ID)
has both cellID and UMI</p></td>
<td><p>All the barcodes(clone IDs) associated with the same cell ID and UMI are replaced by the consensus sequence</p></td>
<td><p>Hamming distance of 5 and all the sequences in a cluster are marked to have the sequuence that is most frequent in that cluster</p></td>
<td><p>Used mutually exclusive genes to detect doublets</p></td>
<td><p>No</p></td>
<td><p>No</p></td>
<td><p>10X, smartSeq-2, smartSeq-3</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>LARRY</p></td>
<td><p>28</p></td>
<td><p>414,811±22,194</p></td>
<td><p>No</p></td>
<td><p>Yes, the amplified library is split into 2 and one of them is used to separately amplify the barcodes</p></td>
<td><p>Abundant UMIs for reads in cells,&lt;20% mitochondrial genes</p></td>
<td><p>minimum of 29 bp of gfp has to be present in the read, the cell ID must be present in the scRNAseq data, 
minimum reads for (cellID-Barcode-UMI) is 10</p></td>
<td><p>Yes and there is cutoff for minimum number of reads(10)</p></td>
<td><p>Hamming distance of 3 - graph connected components based algorithm</p></td>
<td><p>Scrublet- doublet score thresholded based on manual inspection</p></td>
<td><p>Mostly yes</p></td>
<td><p>No</p></td>
<td><p>inDrops, 10X</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>SPLINTR</p></td>
<td><p>60</p></td>
<td><p>3 library sizes - 2 × 10e5, 8 × 10e5 and 1.2 × 10e6</p></td>
<td><p>They can be separately sequenced and the Barcode is identified from the barcode based on the sequences flanking the barcode and then filtered based on some metrics: correct upstream constant region, PHRED score &gt; 30 average, did not contain N residues, had expected barcode pattern - then it was mapped to SPLINTR reference library with 2 mismatches</p></td>
<td><p>No (it is captured along with scRNAseq)</p></td>
<td><p>Filtered to remove cells with &lt;500 genes, &lt;1000 UMIs, &gt;15% mitichondrial gene content</p></td>
<td><p>The barcodes are identified based on the sequences flanking it, have correct upstream sequence, had correct pattern, did not have NN sequences, have PHRED quality &gt;30 all over the barcode, mapped to the reference index upto 2 mismatches</p></td>
<td><p>Yes they are collapsed into 1</p></td>
<td><p>Starcode was used for creating a list of barcodes in library but not otherwsie</p></td>
<td><p>Yes, based on hashtag oligo combinations with the HTODemux function. Also,  droplets containing a combination of two or more SPLINTR barcodes not observed in any other cell were considered to be doublets and were removed.</p></td>
<td><p>Yes</p></td>
<td><p>yes</p></td>
<td><p>10X</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Watermelon</p></td>
<td><p>30</p></td>
<td><p>5,472,944 unique lineage barcodes detected in the Watermelon library</p></td>
<td><p>Yes,the whole transciptome obtained is amplified, targetted lbraries are gel separated and then amplified</p></td>
<td><p>Yes</p></td>
<td><p>Cells should have between 1000-4200 genes, less than 10% mitichondrial gene content</p></td>
<td><p>Need to have the necessary barcode pattern, the length should be exactly 30 and the cell IDs containing these barcodes must be seen in 10X data also</p></td>
<td><p>Yes, they are collapsed by using only unique entries of CellID-Barcode_UMI sequence</p></td>
<td><p>Yes, they have a method:
1) Count number of UMIs supporting each barcode
2) Find the edit distance to the most common barcodes for all lineage barcodes
3)If edit distance is less than 3, add the UMIs of this barcode to the most common barcode and remove this entry
4) If a barcode associated with the cell had UMI count less than 3 times the UMI count of the barcode with max UMI count, this barcode was removed</p></td>
<td><p>No</p></td>
<td><p>Maybe?</p></td>
<td><p>Yes, but not in the original paper</p></td>
<td><p>10X</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>CellTag-Multi</p></td>
<td><p>Depends on library - 28 or 20</p></td>
<td><p>Many counts: 80000, 138867</p></td>
<td><p>no</p></td>
<td><p>The cDNA obtained during 10X sequencing is amplified; after scATAC seq, it is amplified using primers</p></td>
<td><p>Low-quality cells with high mitochondrial reads, low UMIs and features per cell were removed,</p></td>
<td><p>Cells with 1-25 barcodes are kept. Cell ID- CellTag - UMI supported by just 1 read are removed</p></td>
<td><p>Yes</p></td>
<td><p>Starcode and recommended distance of 4</p></td>
<td><p>No</p></td>
<td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p>10X, Drop-Seq</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>ClonMapper</p></td>
<td><p>20</p></td>
<td><p>~7.6 * 10e7 unique barcodes</p></td>
<td><p>Yes, it was done for both pre- and post-treated cells</p></td>
<td><p>No, it was directly read from the BAM file after CellRanger with no amplification step</p></td>
<td><p>Minimum transcript counts - 4000, percentage content of mitochondrial genes - 8%, minimum number of genes sequenced - 750 (the exact values depends on dataset)</p></td>
<td><p>The 3’ and 5’ adapters flanking the barcode are used to identify the region, minimum - 3’ adapter + 10 bp barcode sequence. Also, for gDNA sequenced barcodes, singleton barcodes that were not clustered in Starcode were discarded.</p></td>
<td><p>For barcode reads less than 20, all barcode reads within the cell were compared. If an exact match was found for the 3’ end of a longer barcode, this read was extended. These corrected barcodes were omitted if they were absent in the list of known clones in TP0. The UMI-replicate reads were resolved by using the most frequent barcode read for each cell if none of the other barcodes in the cell had a count greater than 50% of the most frequent barcode. If this was not true, the clonal reads were all removed from the cell</p></td>
<td><p>Yes for gDNA sequencing of cells, starcode is used implementing a message-passing algorithm with a maximum Levenshtein distance of 1 and minimum radius size of 3.</p></td>
<td><p>No</p></td>
<td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p>10X, Illumina</p></td>
<td></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="about.html" class="btn btn-neutral float-left" title="About" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Goyal Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>