<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>singletCode 1.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/sphinx_rtd_size.css" type="text/css" />
      <link rel="stylesheet" href="_static/colour.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script type="text/javascript">document.addEventListener('DOMContentLoaded', function() {document.querySelectorAll('a.reference.external').forEach(function(a) {a.setAttribute('target', '_blank'); a.setAttribute('rel', 'noopener noreferrer');});});</script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            singletCode
          </a>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-doubletInformation">doublets in sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-singletCode">what is singletCode?</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-classifier">classifier</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-usageInformation">implementing singletCode on your data</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-doubletDetectionMethods">benchmarking doublet detection methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-datasetsInfo">dataset information</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-packageDocumentation">singletCode PyPI package</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-infoCLI">command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-watermelonDatasetVignette">Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-singletCodePackageVignette">Vignette to use singletCode package</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-Contact">questions? contact us!</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.cell.com/cell-genomics/fulltext/S2666-979X(24)00176-9">link to paper</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.goyallab.org/">Goyal Lab website</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-repositories">repositories</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">singletCode</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">singletCode 1.2 documentation</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="welcome-to-the-documentation-of-singletcode">
<h1>Welcome to the documentation of singletCode!<a class="headerlink" href="#welcome-to-the-documentation-of-singletcode" title="Permalink to this heading">¶</a></h1>
<p>singletCode <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2023.08.04.552078v2">(link to preprint)</a> uses cell barcoding technology to identify <strong>true singlets</strong> in scRNAseq data. In our preprint, we used singletCode to benchmark existing doublet detection methods and train a classifier to detect singlets in non-barcoded datasets as well of the same cell type.</p>
<div class="toctree-wrapper compound">
<span id="document-doubletInformation"></span><section id="doublets-in-sequencing">
<h2>doublets in sequencing<a class="headerlink" href="#doublets-in-sequencing" title="Permalink to this heading">¶</a></h2>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#how-do-doublets-form" id="id4">How do doublets form?</a></p></li>
<li><p><a class="reference internal" href="#downstream-impact-of-doublets" id="id5">Downstream impact of doublets</a></p></li>
</ul>
</div>
<section id="how-do-doublets-form">
<h3><a class="toc-backref" href="#id4">How do doublets form?</a><a class="headerlink" href="#how-do-doublets-form" title="Permalink to this heading">¶</a></h3>
<p>Single-cell RNA sequencing (scRNA-seq) datasets contain true single cells, or singlets, in addition to cells that randomly coalesce during the protocol, or doublets. Sometimes, there are higher rates of doublets which can be attributed to cellular physiology and experimental protocols can lead to cell clumping. Doublet percentage in a sample be as high as 40%. Doublets can be two very transcriptionally different cells captured together (heterotypic) or two transcriptionally similar cells captured together (homotypic). Doublets are difficult to identify because just because two cells are captured together does not mean there is simply more absolute RNA fragments present or sequenced in doublet cases, and cells exist on a transcriptional continuum, making identification of valid singlet cells difficult, especially if cells are in a transitioning or reprogrammed state.</p>
<figure class="align-default" id="id1">
<a class="reference internal image-reference" href="_images/Figure1A.png"><img alt="How doublets are generated during sequencing" src="_images/Figure1A.png" style="width: 562.0px; height: 265.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of how cells with lineage barcodes appear in the single-cell sequencer where droplets add a unique cell ID. The singlet depicted is one reaction droplet with a single cell such that there is a 1:1:1 mapping of a cell to a barcode to a cell ID. Alternatively, an example of a doublet is when one droplet has two or more cells, each having a unique barcode.</em></span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="downstream-impact-of-doublets">
<h3><a class="toc-backref" href="#id5">Downstream impact of doublets</a><a class="headerlink" href="#downstream-impact-of-doublets" title="Permalink to this heading">¶</a></h3>
<p>Doublets are problematic because they can impact the conclusions of scRNA-seq downstream functional analyses. We found that doublets confound downstream analyses when we tested common downstream scRNA-seq data analysis protocols (differential expression, cell trajectory, clustering stability, and cell-cell communication) on datasets of various doublet percentages. A schematic for how these datasets were generated and examples of our results for differential expression are below.</p>
<figure class="align-center" id="id2">
<a class="reference internal image-reference" href="_images/Figure5A.png"><img alt="Schematic of datasets generated for all functional analyses. The dataset is first processed to have 40% doublets and undergo standard Seurat processing with Leiden clustering, where cluster labels here are represented in different colors in UMAP space (second box, left), and doublet labels are represented by different shades of blue in UMAP space (second box, right). Then,doublets are removed to achieve 20%, 10% and 0% doublet datasets. The 0%, 10%, 20% and 40% doublet rate datasets are generated for each dataset used in benchmarking except for Smart-seq3 to evaluate doublet impact on various downstream analyses: cell trajectory, cell-cell communication, clustering stability, and differential expression." src="_images/Figure5A.png" style="width: 309.5px; height: 489.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of datasets generated for all functional analyses. The dataset is first processed to have 40% doublets and undergo standard Seurat processing with Leiden clustering, where cluster labels here are represented in different colors in UMAP space (second box, left), and doublet labels are represented by different shades of blue in UMAP space (second box, right). Then,doublets are removed to achieve 20%, 10% and 0% doublet datasets. The 0%, 10%, 20% and 40% doublet rate datasets are generated for each dataset used in benchmarking except for Smart-seq3 to evaluate doublet impact on various downstream analyses: cell trajectory, cell-cell communication, clustering stability, and differential expression.</em></span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<div class="line-block">
<div class="line"><strong>differential expression</strong></div>
<div class="line">Differential expression is used to identify the genes that are differentially expressed between cells belonging to different conditions. The genes identified could be used to explain the phenotype observed or examine pathways involved. Our results indicated that increasing doublet rate decreases the precision, recall, and true negative rate (TNR) for the datasets when compared to the differentially expressed genes inferred from the clean singlets-only dataset, with a higher doublet rate leading to worse performance.</div>
</div>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="_images/Figure5B.png"><img alt="Differential expression analysis results for all datasets using (blue, right) on two randomly chosen clusters at various doublet rates. Line color shade represents the % doublet rate, with a more saturated color indicative of a higher doublet rate. Precision, recall and TNR are calculated against ground truth differentially expressed genes in the corresponding dataset without doublets. P-values between variable doublet rate datasets are calculated with a paired Wilcoxon signed-rank test." src="_images/Figure5B.png" style="width: 536.5px; height: 511.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Differential expression analysis results for all datasets using MAST (purple, left) and Wilcox (blue, right) on two randomly chosen clusters at various doublet rates. Line color shade represents the % doublet rate, with a more saturated color indicative of a higher doublet rate. Precision, recall and TNR are calculated against ground truth differentially expressed genes in the corresponding dataset without doublets. P-values between variable doublet rate datasets are calculated with a paired Wilcoxon signed-rank test.</em></span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<div class="line-block">
<div class="line"><strong>clustering stability</strong></div>
<div class="line">Clustering is one of the most common analyses done to infer the identity of similar cells. When the number of doublets were increased, the probability of getting the correct number of cell clusters decreased. This might lead to spurious clusters or a genuine sub-type of cells not being identified as a distinct group. For details about our results, see our paper.</div>
<div class="line"><br /></div>
<div class="line"><strong>cell-cell communication</strong></div>
<div class="line">Cell-cell communication inference can provide information about intercellular communication networks. This can be used to understand the role of cell-cell interactions in biological processes. Increasing the doublet rate decreases the precision and recall for communication pathways identified when compared to the clean singlets only datasets. For details about our results, see our paper.</div>
<div class="line"><br /></div>
<div class="line"><strong>cell trajectory</strong></div>
<div class="line">Increasing doublet rate caused the trajectory to deviate away from the trajectory inferred from the clean singlet-only datasets and there were gain/loss of lineages when doublets were introduced. For details about our results, see our paper.</div>
</div>
</section>
</section>
<span id="document-singletCode"></span><section id="what-is-singletcode">
<h2>what is singletCode?<a class="headerlink" href="#what-is-singletcode" title="Permalink to this heading">¶</a></h2>
<p>singletCode is a framework to extract true singlets from barcoded scRNA-seq data. Our pipeline identified barcoded singlets from 10 different publications and 6 original experimental scRNA-seq samples generated in this study, encompassing 7 different barcoding technologies, 94 scRNA-seq samples, 3 unique sequencing technologies, and a total of 564,579 cells. Of the 338,948 barcoded cells, we extracted 293,618 singlets.(read more about the barcoded datasets we used singletCode with <a class="reference internal" href="index.html#datasetinfo"><span class="std std-ref">here</span></a>).</p>
<figure class="align-center">
<a class="reference internal image-reference" href="_images/Figure1B.png"><img alt="Stacked bar chart showing types of singlets in each dataset. Each color represents one type of barcode composition that constitutes a singlet. Single barcode means one cell ID has one associated lineage barcode. Dominant barcode means one barcode has the highest UMI counts in one cell. Multiple barcodes within or across samples are scenarios where the same barcode combinations are observed in the same sample or across twin samples. See our paper for details." src="_images/Figure1B.png" style="width: 711.5px; height: 165.0px;" /></a>
</figure>
<section id="singlet-identification">
<h3>Singlet identification<a class="headerlink" href="#singlet-identification" title="Permalink to this heading">¶</a></h3>
<p>singletCode leverages the fact that the lineage barcodes are present within the cell before sequencing. As a result of multiplicity of infection, singletCode identifies cells which meet any of the following conditions as a true singlet:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>1 barcode/1 cell ID</p></li>
<li><p>&gt;1 barcode per 1 cell ID, but 1 barcode has significantly more UMI counts than the other barcodes within the same cell</p></li>
<li><p>M barcodes per 1 cell ID, but the same combination of M barcodes are found in other cells in the same sample</p></li>
<li><p>M barcodes per 1 cell ID, but the same combination of M barcodes are found in other cells across samples within the same experimental design (common in barcoding studies)</p></li>
</ol>
</div></blockquote>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="_images/Figure1F.png"><img alt="Stacked bar chart showing types of singlets in each dataset. Each color represents one type of barcode composition that constitutes a singlet. Single barcode means one cell ID has one associated lineage barcode. Dominant barcode means one barcode has the highest UMI counts in one cell. Multiple barcodes within or across samples are scenarios where the same barcode combinations are observed in the same sample or across twin samples. See our paper for details." src="_images/Figure1F.png" style="width: 634.0px; height: 373.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Stacked bar chart showing types of singlets in each dataset. Each color represents one type of barcode composition that constitutes a singlet. Single barcode means one cell ID has one associated lineage barcode. Dominant barcode means one barcode has the highest UMI counts in one cell. Multiple barcodes within or across samples are scenarios where the same barcode combinations are observed in the same sample or across twin samples. See our paper for details.</em></span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
</section>
<span id="document-classifier"></span><section id="classifier">
<h2>classifier<a class="headerlink" href="#classifier" title="Permalink to this heading">¶</a></h2>
<section id="how-was-it-built">
<h3>How was it built?<a class="headerlink" href="#how-was-it-built" title="Permalink to this heading">¶</a></h3>
<p>We used the true singlets identified with singletCode to simulate datasets with 10% doublets and trained an extreme gradient-boosting (XGBoost) classifier to detect these doublets.</p>
<p>The classifier is built in two steps:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Hyperparameter optimization (using hyperopt()) with a training and validation set.</p></li>
<li><p>Training the optimal model with a training and test set</p></li>
</ol>
</div></blockquote>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="_images/Figure6A.png"><img alt="Schematic of classifier optimization and training. Data was first split into training, validation, and testing sets. The training and validation sets were used to optimize the hyperparameters using the hyperopt library. A search space (blue cube) was defined in high dimensional hyperparameter space and, through Bayesian optimization, the optimal value for each hyperparameter was determined based on maximizing the AUPRC value. Once the model’s parameters were optimized, the final, optimal, classifier was trained using the training and test sets. An extreme gradient boosted (XGBoost) classifier was trained to recognize doublet cells with input data being gene count matrices of datasets made from singlets and 10% simulated doublets." src="_images/Figure6A.png" style="width: 463.20000000000005px; height: 808.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of classifier optimization and training. Data was first split into training, validation, and testing sets. The training and validation sets were used to optimize the hyperparameters using the hyperopt library. A search space (blue cube) was defined in high dimensional hyperparameter space and, through Bayesian optimization, the optimal value for each hyperparameter was determined based on maximizing the AUPRC value. Once the model’s parameters were optimized, the final, optimal, classifier was trained using the training and test sets. An extreme gradient boosted (XGBoost) classifier was trained to recognize doublet cells with input data being gene count matrices of datasets made from singlets and 10% simulated doublets.</em></span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="how-well-does-it-perform">
<h3>How well does it perform?<a class="headerlink" href="#how-well-does-it-perform" title="Permalink to this heading">¶</a></h3>
<p>We achieved significantly higher AUPRC and AUROC scores using our classifier compared to the other methods we benchmarked for doublet detection.</p>
<figure class="align-center" id="id2">
<a class="reference internal image-reference" href="_images/Figure6B.png"><img alt="Results of classifier doublet detection compared to other doublet detection methods as measured by AUPRC. Each dot represents the average AUPRC score of a doublet detection method (color) on a given dataset (x-axis). The ribbon has the width of the standard deviation of the AUPRC score for each doublet detection method." src="_images/Figure6B.png" style="width: 464.4px; height: 309.3px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Results of classifier doublet detection compared to other doublet detection methods as measured by AUPRC. Each dot represents the average AUPRC score of a doublet detection method (color) on a given dataset (x-axis). The ribbon has the width of the standard deviation of the AUPRC score for each doublet detection method.</em></span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="classifying-doublets-in-non-barcoded-datasets">
<span id="classifier-nonebarcoded"></span><h3>Classifying doublets in non-barcoded datasets<a class="headerlink" href="#classifying-doublets-in-non-barcoded-datasets" title="Permalink to this heading">¶</a></h3>
<p>Although barcoding experiments are becoming increasingly prevalent, they are still relatively uncommon. Therefore, we sought to train a doublet classifier on barcoded data that could detect doublets in non-barcoded data. We trained a classifier on cell samples from melanoma, mouse brain, leukemia, and bone marrow/leukemia using true singlet labels from singletCode and successfully identified doublets in a similar cell sample without needing barcoding data. Further, we integrated all the data from mouse and human samples together, split it in half, and it to train a classifier that was then tested on individual sample components of the opposite half. This, too, proved to outperform other doublet detection methods.</p>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="_images/Figure6G.png"><img alt="Schematic of training a doublet classifier on barcoded data from 1 experiment and using that classifier to identify doublets in a biological replicate, experiment 2." src="_images/Figure6G.png" style="width: 253.0px; height: 329.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of training a doublet classifier on barcoded data from 1 experiment and using that classifier to identify doublets in a biological replicate, experiment 2.</em></span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>As barcoded scRNA-seq data becomes more abundant, experimenters can train a classifier specific to their cell type using a barcoded subset of data, and, eventually, there could be enough data to train a classifier on multiple barcoded cell types which can be used more generally.</p>
</section>
</section>
<span id="document-usageInformation"></span><section id="implementing-singletcode-on-your-data">
<h2>implementing singletCode on your data<a class="headerlink" href="#implementing-singletcode-on-your-data" title="Permalink to this heading">¶</a></h2>
<p>singletCode is a framework that can be extended to most scRNA-seq or even scATAC-seq data to accurately identify singlets.</p>
<section id="barcoded-datasets">
<h3>Barcoded datasets<a class="headerlink" href="#barcoded-datasets" title="Permalink to this heading">¶</a></h3>
<p>We created both a python package called singletCode and a command-line interface. The only input needed is a .csv file with cell ID, barcode, sample information, and UMI counts for each barcode. You can find more information to use it <a class="reference internal" href="index.html#package-info"><span class="std std-ref">here</span></a></p>
</section>
<section id="non-barcoded-datasets">
<h3>Non-barcoded datasets<a class="headerlink" href="#non-barcoded-datasets" title="Permalink to this heading">¶</a></h3>
<p>If the data is non-barcoded, the information from true singlets identified from singletCode can still be harnessed using a classifier trained on barcoded data from a different experiment using the cell type of interest. Find more information about this in the classifier page <a class="reference internal" href="index.html#classifier-nonebarcoded"><span class="std std-ref">here</span></a>.</p>
</section>
</section>
<span id="document-doubletDetectionMethods"></span><section id="benchmarking-doublet-detection-methods">
<h2>benchmarking doublet detection methods<a class="headerlink" href="#benchmarking-doublet-detection-methods" title="Permalink to this heading">¶</a></h2>
<section id="simulation-datasets">
<h3>Simulation datasets<a class="headerlink" href="#simulation-datasets" title="Permalink to this heading">¶</a></h3>
<p>To simulate doublets for benchmarking, we randomly selected the gene expression counts data from two cells that were found to be true singlets by singletCode. We averaged the counts from these two cells to generate simulated doublets and create datasets with various doublet percentages for benchmarking.</p>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="_images/Figure1C.png"><img alt="Workflow schematic of how raw data is processed to generate true singlets based on a barcode UMI cutoff and doublet removal based on singletCode specifications, followed by the simulation of doublets for benchmarking doublet detection methods." src="_images/Figure1C.png" style="width: 213.5px; height: 389.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Workflow schematic of how raw data is processed to generate true singlets based on a barcode UMI cutoff and doublet removal based on singletCode specifications, followed by the simulation of doublets for benchmarking doublet detection methods.</em></span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="benchmarking">
<h3>Benchmarking<a class="headerlink" href="#benchmarking" title="Permalink to this heading">¶</a></h3>
<dl class="simple">
<dt>We used these datasets to benchmark four doublet detection methods:</dt><dd><ol class="arabic simple">
<li><p><a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/scDblFinder.html">scDblFinder</a>,</p></li>
<li><p><a class="reference external" href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a>,</p></li>
<li><p><a class="reference external" href="https://github.com/AllonKleinLab/scrublet/">Scrublet</a> and</p></li>
<li><p><a class="reference external" href="https://github.com/kostkalab/scds">Hybrid</a></p></li>
</ol>
</dd>
</dl>
<p>We evaluated the AUPRC, AUROC, TNR, and doublet scores and calls of the four methods and found lower than expected performance for all methods. A plot of our results for AUPRC value is found below.</p>
<figure class="align-center" id="id2">
<a class="reference internal image-reference" href="_images/Figure2A.png"><img alt="Color-coded boxplots of AUPRC value for all four doublet detection methods. Each boxplot is calculated from the AUPRC value after running the respective method on each sample of the dataset with actual doublet rate of 0.08 and expected doublet rate set to 0.05, 0.08, 0.1, 0.15, 0.2, and 0.25 where applicable. Dots represent the mean AUPRC value for a detection method grouped across all samples within a dataset. Lines represent standard deviation of the AUPRC values for a detection method across all samples within a dataset. Boxes span the first and third quartiles. Outliers are not depicted." src="_images/Figure2A.png" style="width: 634.8000000000001px; height: 221.60000000000002px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Color-coded boxplots of AUPRC value for all four doublet detection methods. Each boxplot is calculated from the AUPRC value after running the respective method on each sample of the dataset with actual doublet rate of 0.08 and expected doublet rate set to 0.05, 0.08, 0.1, 0.15, 0.2, and 0.25 where applicable. Dots represent the mean AUPRC value for a detection method grouped across all samples within a dataset. Lines represent standard deviation of the AUPRC values for a detection method across all samples within a dataset. Boxes span the first and third quartiles. Outliers are not depicted.</em></span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>We examined the consistency of doublet labeling across different doublet detection methods by introducing a ‘similarity score’—a measure of the fraction of doublets identically classified by two methods. With an average similarity score of 0.66 across all datasets and methods, we observed variability in doublet detection.</p>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="_images/Figure4A.png"><img alt="Pairwise similarity score between all benchmarked methods. A score of 1 indicates perfect identity and a 0 indicates complete disagreement. Average similarity score across all methods and datasets is 0.66." src="_images/Figure4A.png" style="width: 670.5px; height: 492.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Pairwise similarity score between all benchmarked methods. A score of 1 indicates perfect identity and a 0 indicates complete disagreement. Average similarity score across all methods and datasets is 0.66.</em></span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>We further evaluated doublet detection on ensemble doublet detection methods (hybrid, Chord) and across sequencing technologies (10X Genomics, Smart-seq3). For a more detailed evaluation of our results, refer to our paper.</p>
</section>
<section id="heterogeneity-effects">
<h3>Heterogeneity effects<a class="headerlink" href="#heterogeneity-effects" title="Permalink to this heading">¶</a></h3>
<p>We wanted to know whether heterogeneity of a dataset affects the performance of doublet detection methods. Because heterogeneity can be impacted by many properties of a dataset, such as experimental design and data processing, we made conclusions based on heterogeneity within a sample. We did this by subsampling singlets and doublets within a single PC cluster for a sample (less heterogeneous, low Euclidean distance), and across all clusters for a sample (more heterogeneity, higher Euclidean distance).</p>
<figure class="align-default" id="id4">
<a class="reference internal image-reference" href="_images/Figure3B.png"><img alt="Schematic of how transcriptionally similar and dissimilar cells are subsampled for subsequent within-sample heterogeneity determinations. A principal component analysis (left) was performed on each dataset and the first 30 PCs are used to identify Louvain clusters, represented by color. Transcriptionally similar (less heterogeneous) cells are identified from a single cluster and have a low Euclidean distance in PC space. Transcriptionally dissimilar (more heterogeneous) cells are identified from all clusters within the dataset and have a high average Euclidean distance in PC space." src="_images/Figure3B.png" style="width: 766.5px; height: 171.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of how transcriptionally similar and dissimilar cells are subsampled for subsequent within-sample heterogeneity determinations. A principal component analysis (left) was performed on each dataset and the first 30 PCs are used to identify Louvain clusters, represented by color. Transcriptionally similar (less heterogeneous) cells are identified from a single cluster and have a low Euclidean distance in PC space. Transcriptionally dissimilar (more heterogeneous) cells are identified from all clusters within the dataset and have a high average Euclidean distance in PC space.</em></span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="singletcode-for-scatac-seq-doublet-detection">
<h3>singletCode for scATAC-seq doublet detection<a class="headerlink" href="#singletcode-for-scatac-seq-doublet-detection" title="Permalink to this heading">¶</a></h3>
<p>We evaluated our method’s ability to assess doublet detection in scATAC-seq datasets using AMULET for a proof-of-concept analysis. We generated Watermelon-barcoded 10X Genomics Multiome datasets and applied AMULET to scATAC-seq fragments to categorize singlets and doublets. Concurrently, we identified true singlets in a barcoded scRNA-seq library from the same cells. By comparing the true negative and false positive rates between AMULET and singletCode across six datasets, we calculated AMULET’s average true negative rate (TNR) at 0.924. Our results demonstrate that singletCode can be used to benchmark doublet detection in other modalities besides scRNA-seq.</p>
<figure class="align-center" id="id5">
<a class="reference internal image-reference" href="_images/Figure4G.png"><img alt="Schematic for 10X Single Cell Multiome dataset workflow resulting in TNR calculation. The same cells in the Multiome dataset undergo scATAC-seq, scRNA-seq, and barcode sequencing. Barcodes were processed with singletCode to identify ground truth singlets while the respective fragments file from ATAC was used to label doublets with AMULET. The AMULET-labeled singlets were compared with ground truth singletCode singlets to calculate TNR." src="_images/Figure4G.png" style="width: 323.0px; height: 383.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic for 10X Single Cell Multiome dataset workflow resulting in TNR calculation. The same cells in the Multiome dataset undergo scATAC-seq, scRNA-seq, and barcode sequencing. Barcodes were processed with singletCode to identify ground truth singlets while the respective fragments file from ATAC was used to label doublets with AMULET. The AMULET-labeled singlets were compared with ground truth singletCode singlets to calculate TNR.</em></span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
</section>
<span id="document-datasetsInfo"></span><section id="dataset-information">
<span id="datasetinfo"></span><h2>dataset information<a class="headerlink" href="#dataset-information" title="Permalink to this heading">¶</a></h2>
<section id="barcoding-technologies-assessed">
<h3>barcoding technologies assessed<a class="headerlink" href="#barcoding-technologies-assessed" title="Permalink to this heading">¶</a></h3>
<p>7 different barcoding technologies
We incorporated datasets using FateMap, ClonMapper, SPLINTR, LARRY, CellTag-multi, Watermelon, and TREX barcodes. Each have their own recovery and analysis specifications that we outline in Box 1 of our paper.</p>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="_images/Box1.png"><img alt="Datasets from 10 different publications spanning dozens of cell types." src="_images/Box1.png" style="width: 573.5px; height: 436.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of the structure of different lineage-tracing barcodes, their length, diversity, and fluorescent color. Each barcode contains an expressed fluorescent protein, a unique sequence, known as the barcode, and a polyA tail for capture. The lengths of barcodes in each method are shown as scale bars, and library diversities are shown as circles, sized proportional to their diversity and colored according to fluorescent proteins used.</em></span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="dataset-details">
<h3>dataset details<a class="headerlink" href="#dataset-details" title="Permalink to this heading">¶</a></h3>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Specifications for barcoded and non-barcoded datasets</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 8%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Dataset Name</p></th>
<th class="head"><p>Alias ID</p></th>
<th class="head"><p>Sample type</p></th>
<th class="head"><p>Organism</p></th>
<th class="head"><p>Tissue</p></th>
<th class="head"><p>Cell line</p></th>
<th class="head"><p>Genotype</p></th>
<th class="head"><p>Treatment</p></th>
<th class="head"><p>Total samples</p></th>
<th class="head"><p>Publication</p></th>
<th class="head"></th>
<th class="head"></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td></td>
<td><p>barcoded scRNA-seq datasets</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Goyal et al. 1</p></td>
<td><p>FM01</p></td>
<td><p>Melanoma patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Skin-derived</p></td>
<td><p>WM989 A6-G3</p></td>
<td><p>BRAF – V600E</p></td>
<td><p>Targeted therapy</p></td>
<td><p>4</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Goyal et al. 2</p></td>
<td><p>FM02</p></td>
<td><p>Melanoma patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Skin-derived</p></td>
<td><p>WM989 A6-G3</p></td>
<td><p>BRAF – V600E</p></td>
<td><p>Targeted therapy</p></td>
<td><p>4</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Goyal et al. 3</p></td>
<td><p>FM03</p></td>
<td><p>Melanoma patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Skin-derived</p></td>
<td><p>WM989 A6-G3</p></td>
<td><p>BRAF – V600E</p></td>
<td><p>Targeted therapy</p></td>
<td><p>2</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Goyal et al. 4</p></td>
<td><p>FM04</p></td>
<td><p>Breast cancer patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Breast; mammary gland</p></td>
<td><p>MDA-MB-231-D4</p></td>
<td><p>BRAF – G464V, KRAS – G13D, TERT – c.1-124C&gt;T, TP53 – R280K</p></td>
<td><p>Chemotherapy, cytotoxic</p></td>
<td><p>2</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Goyal et al. 5</p></td>
<td><p>FM05</p></td>
<td><p>Melanoma patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Skin-derived</p></td>
<td><p>WM989 A6-G3</p></td>
<td><p>BRAF – V600E</p></td>
<td><p>Targeted therapy</p></td>
<td><p>2</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Goyal et al. 6</p></td>
<td><p>FM06</p></td>
<td><p>Melanoma patient untreated cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Skin-derived</p></td>
<td><p>WM989 A6-G3</p></td>
<td><p>BRAF – V600E</p></td>
<td><p>No treatment</p></td>
<td><p>2</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Goyal et al. 8</p></td>
<td><p>FM08</p></td>
<td><p>Primary melanocytes</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Skin-derived</p></td>
<td><p>FOM230-1</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>2</p></td>
<td><p>Goyal et al., Nature, 2023 (ref. 28)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Jiang et al.</p></td>
<td><p>non_cancer</p></td>
<td><p>HIPS differentiation</p></td>
<td><p>Homo sapiens</p></td>
<td><p>PBMCs</p></td>
<td><p>PENN123i-SV20</p></td>
<td><p>WT</p></td>
<td><p>Cardiac differentiation signal</p></td>
<td><p>2</p></td>
<td><p>Jiang et al., Genome Biol, 2022 (ref. 31)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Jain et al.</p></td>
<td><p>Biorxiv</p></td>
<td><p>Stem cell reprogramming</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Fibroblasts</p></td>
<td><p>hiF-T</p></td>
<td><p>WT</p></td>
<td><p>OKSM</p></td>
<td><p>6</p></td>
<td><p>Jain et al., Cell Systems, 2024 (ref. 34)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>TREX</p></td>
<td><p>TREX</p></td>
<td><p>neuroepithelial progenitor cells differentiation</p></td>
<td><p>Mus musculus</p></td>
<td><p>Neural</p></td>
<td><p>CD-1</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>No treatment</p></td>
<td><p>4</p></td>
<td><p>Ratz et al., Nature neuroscience, 2022 (ref. 37)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>LARRY</p></td>
<td><p>LARRY, LK</p></td>
<td><p>pan-myeloid differentiation</p></td>
<td><p>Mus musculus</p></td>
<td><p>Blood</p></td>
<td><p>Lin-Kit+</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>cytokines and growth factors</p></td>
<td><p>18</p></td>
<td><p>Weinreb et al., Science, 2020 (ref. 35)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>LARRY</p></td>
<td><p>LARRY, LSK</p></td>
<td><p>pan-myeloid differentiation</p></td>
<td><p>Mus musculus</p></td>
<td><p>Blood</p></td>
<td><p>Lin-Kit+Sca-1+</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>cytokines and growth factors</p></td>
<td><p>15</p></td>
<td><p>Weinreb et al., Science, 2020 (ref. 35)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>SPLINTR</p></td>
<td><p>SPLINTR_chemo</p></td>
<td><p>Acute myeloid leukaemia</p></td>
<td><p>Mus musculus</p></td>
<td><p>Bone marrow</p></td>
<td><p>C57BL/6</p></td>
<td><p>MLL-AF9 + KrasG12D</p></td>
<td><p>chemotherapy</p></td>
<td><p>10</p></td>
<td><p>Fennell et al., Nature, 2022 (ref. 26)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>SPLINTR</p></td>
<td><p>SPLINTR_clone</p></td>
<td><p>Acute myeloid leukaemia</p></td>
<td><p>Mus musculus</p></td>
<td><p>Bone marrow</p></td>
<td><p>C57BL/6</p></td>
<td><p>MLL-AF9, MLL-AF9 + KrasG12D, MLL-AF9 + Flt3ITD</p></td>
<td><p>No treatment</p></td>
<td><p>4</p></td>
<td><p>Fennell et al., Nature, 2022 (ref. 26)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>SPLINTR</p></td>
<td><p>SPLINTR_chemoretrans</p></td>
<td><p>Acute myeloid leukaemia</p></td>
<td><p>Mus musculus</p></td>
<td><p>Bone marrow and spleen</p></td>
<td><p>C57BL/6</p></td>
<td><p>MLL-AF9 + KrasG12D</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Fennell et al., Nature, 2022 (ref. 26)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Watermelon</p></td>
<td><p>Watermelon</p></td>
<td><p>Breast cancer patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Breast; mammary gland</p></td>
<td><p>T-47D</p></td>
<td><p>PIK3CA – H1047R; 
Tp53 – L194F</p></td>
<td><p>Targeted therapy</p></td>
<td><p>6</p></td>
<td><p>This paper</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>CellTag-multi</p></td>
<td><p>CellTag-multi_d</p></td>
<td><p>multilineage differentiation (hematopoiesis)</p></td>
<td><p>Mus musculus</p></td>
<td><p>bone marrow</p></td>
<td><p>LSK</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>cytokines and growth factors</p></td>
<td><p>3</p></td>
<td><p>Jindal et al., Nature biotechnology, 2023 (ref. 36)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>CellTag-multi</p></td>
<td><p>CellTag-multi_B4</p></td>
<td><p>iEP reprogramming</p></td>
<td><p>Mus musculus</p></td>
<td><p>embryo</p></td>
<td><p>C57BL/6J</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>EGF</p></td>
<td><p>1</p></td>
<td><p>Jindal et al., Nature biotechnology, 2023 (ref. 36)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>ClonMapper</p></td>
<td><p>ClonMapper</p></td>
<td><p>CLL resistant cell lines</p></td>
<td><p>Homo sapiens</p></td>
<td><p>blood</p></td>
<td><p>HG3</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>Chemotherapy</p></td>
<td><p>4</p></td>
<td><p>Gutierrez et al., Nature Cancer, 2021 (ref. 17)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Smart-seq3</p></td>
<td><p>Smart-seq3</p></td>
<td><p>neuroepithelial progenitor cells differentiation</p></td>
<td><p>Mus musculus</p></td>
<td><p>Neural</p></td>
<td><p>CD-1</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>No treatment</p></td>
<td><p>2</p></td>
<td><p>Mold et al., Cell systems, 2024 (ref. 38)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td></td>
<td><p>nonbarcoded sc-RNA seq datasets</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>non-barcoded</p></td>
<td><p>hm-12k</p></td>
<td><p>Synthetic dataset</p></td>
<td><p>Homo sapiens, Mus musculus</p></td>
<td><p>Human kidney, Murine fibroblasts</p></td>
<td><p>HEK293T, NIH3T3</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Zheng et al., Nat Commun, 2017 (ref. 44)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>non-barcoded</p></td>
<td><p>hm-6k</p></td>
<td><p>Synthetic dataset</p></td>
<td><p>Homo sapiens, Mus musculus</p></td>
<td><p>Human kidney, Murine fibroblasts</p></td>
<td><p>HEK293T, NIH3T3</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Zheng et al., Nat Commun, 2017 (ref. 44)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>non-barcoded</p></td>
<td><p>HMEC-orig-MULTI</p></td>
<td><p>HMECs</p></td>
<td><p>Homo sapiens</p></td>
<td><p>mammary gland</p></td>
<td><p>HMEC</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>McGinnis et al., Cell Syst., 2019 (ref. 6)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>non-barcoded</p></td>
<td><p>HMEC-rep-MULTI</p></td>
<td><p>HMECs</p></td>
<td><p>Homo sapiens</p></td>
<td><p>mammary gland</p></td>
<td><p>HMEC</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>McGinnis et al., Cell Syst., 2019 (ref. 6)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>non-barcoded</p></td>
<td><p>HEK-HMEC-MULTI</p></td>
<td><p>Synthetic dataset</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Human kidney, mammary gland</p></td>
<td><p>HEK293T, HMEC</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>McGinnis et al., Cell Syst., 2019 (ref. 6)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>non-barcoded</p></td>
<td><p>mkidney-ch</p></td>
<td><p>Mouse kidney cells</p></td>
<td><p>Mus musculus</p></td>
<td><p>Mouse kidney</p></td>
<td><p>C57BL/6J</p></td>
<td><p>WT</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Bernstein et al., Cell Syst., 2020 (ref. 3)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>non-barcoded</p></td>
<td><p>pbmc-2ctrl-dm</p></td>
<td><p>Patient PBMCs</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Systemic lupus erythematosus (SLE) PBMCs</p></td>
<td><p>Patient- derived</p></td>
<td><p>Patient- derived</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Kang et al., Nat. Biotechnol., 2018 (ref. 45)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>non-barcoded</p></td>
<td><p>pbmc-2stim-dm</p></td>
<td><p>Patient PBMCs</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Systemic lupus erythematosus (SLE) PBMCs</p></td>
<td><p>Patient- derived</p></td>
<td><p>Patient- derived</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Kang et al., Nat. Biotechnol., 2018 (ref. 45)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>non-barcoded</p></td>
<td><p>cline-ch</p></td>
<td><p>Human HEK, K562, KG1, and THP1</p></td>
<td><p>Homo sapiens</p></td>
<td><p>blood</p></td>
<td><p>human - derived</p></td>
<td><p>human - derived</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Stoeckius et al., Genome Biol., 2018 (ref. 13)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>non-barcoded</p></td>
<td><p>pbmc-ch</p></td>
<td><p>Human PBMCs</p></td>
<td><p>Homo sapiens</p></td>
<td><p>blood</p></td>
<td><p>human - derived</p></td>
<td><p>human - derived</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>Stoeckius et al., Genome Biol., 2018 (ref. 13)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>non-barcoded</p></td>
<td><p>pdx-MULTI</p></td>
<td><p>Synthetic dataset</p></td>
<td><p>Homo sapiens, Mus musculus</p></td>
<td><p>Human breast cancer, mouse immue</p></td>
<td><p>PDX mouse model</p></td>
<td><p>PDX mouse model</p></td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>McGinnis et al., Cell Syst., 2019 (ref. 6)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>non-barcoded</p></td>
<td><p>nuc-MULTI</p></td>
<td><p>Synthetic dataset</p></td>
<td><p>Homo sapiens, Mus musculus</p></td>
<td><p>nuclei</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>No treatment</p></td>
<td><p>1</p></td>
<td><p>McGinnis et al., Cell Syst., 2019 (ref. 6)</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td></td>
<td><p>Multiome datasets</p></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>ATAC</p></td>
<td><p>Watermelon Multiome</p></td>
<td><p>Breast cancer patient resistant cell line</p></td>
<td><p>Homo sapiens</p></td>
<td><p>Breast; mammary gland</p></td>
<td><p>MCF7</p></td>
<td><p>GATA3 – D336G; 
PIK3CA – E545K</p></td>
<td><p>Targeted therapy</p></td>
<td><p>6</p></td>
<td><p>This paper</p></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="cell-types">
<h3>cell types<a class="headerlink" href="#cell-types" title="Permalink to this heading">¶</a></h3>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="_images/Figure1D.png"><img alt="Stacked bar chart of cell types in all datasets. Each distinct color corresponds to one study, with the proportions of different cell types arranged from most to least prevalent (deepest to lightest) and separated by vertical lines. Adjacent to the bar chart, sketched tissues and cells provide a visualization of the cell types used in the particular study." src="_images/Figure1D.png" style="width: 472.5px; height: 375.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Stacked bar chart of cell types in all datasets. Each distinct color corresponds to one study, with the proportions of different cell types arranged from most to least prevalent (deepest to lightest) and separated by vertical lines. Adjacent to the bar chart, sketched tissues and cells provide a visualization of the cell types used in the particular study.</em></span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
</section>
<span id="document-packageDocumentation"></span><section id="singletcode-pypi-package">
<h2>singletCode PyPI package<a class="headerlink" href="#singletcode-pypi-package" title="Permalink to this heading">¶</a></h2>
<section id="availablility">
<h3>Availablility<a class="headerlink" href="#availablility" title="Permalink to this heading">¶</a></h3>
<p>The package is available on PyPI as <a class="reference external" href="https://pypi.org/project/singletCode/">singletCode</a> .</p>
</section>
<section id="how-do-i-use-it">
<span id="package-info"></span><h3>How do I use it?<a class="headerlink" href="#how-do-i-use-it" title="Permalink to this heading">¶</a></h3>
<blockquote>
<div><ol class="arabic">
<li><p>Installation</p>
<blockquote>
<div><p>It can be installed from PyPI using the following in the terminal:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pip3</span> <span class="n">install</span> <span class="n">singletCode</span>
<span class="kn">from</span> <span class="nn">singletCode</span> <span class="kn">import</span> <span class="n">check_sample_sheet</span><span class="p">,</span> <span class="n">get_singlets</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Preparing the input sample sheet.</p>
<blockquote>
<div><p>The input sample sheet is a .csv file that contains the information about cell ID (added while sequencing), lineage barcode, and sample name. Each row should be repeated n times where n is the number of UMIs associated with that barcode and cell ID combination. For creating the input, you can read the .csv file in as a pandas dataframe.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;path/to/csv/file.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
<p>You can check if the format and the column names are valid for running singletCode using check_sample_sheet() function.</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">check_sample_sheet</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<p>It will either return an error with information about how to modify the sample sheet to make it a valid input or will print the message, “The sample sheet provided can be used as input to get_singlets to get a list of singlets identified”, in which case you can move to the next step.</p>
</div></blockquote>
</li>
<li><p>Running get_singlets() to get an assignment of singlet status for each cell ID and barcode combination.</p>
<blockquote>
<div><p>This is the step where singlet identification is done using the singletCode framework. You can read more about the parameters for get_singlets() below. In this example, default values are used.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cellLabelList</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="n">get_singlets</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">dataset_name</span> <span class="o">=</span> <span class="s2">&quot;Sample1&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>cellLabelList</strong> is a pandas dataframe that contains 5 rows: cellID, barcode, sample, nUMI and label. Label is whether the particular cell ID and barcode combination has been called a singlet or not.</p>
<p>To get a dataframe containing just the singlets, you can run this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">singletList</span> <span class="o">=</span> <span class="n">cellLabelList</span><span class="p">[</span><span class="n">cellLabelList</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Singlet&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p><strong>stats</strong> contains the statistics for each sample present in your dataset: total cells, total number of singlets, number of singlets recovered from different categories of singlets (such as single-barcode singlets, muli-barcode singlets, dominant-UMI singlets), number of cells removed due to low barcode UMI counts for the barcode and number of indeterminate cells since singletCode can identify only truly singlet cells but not be certain if the other cells are truly not singlets.</p>
<p>You can save the singlets list in a .txt file and the list of cell IDs in each category of singlets by setting <strong>save_all_singlet_categories</strong> parameter to TRUE.</p>
</div></blockquote>
</li>
</ol>
</div></blockquote>
</section>
<section id="detailed-information-about-the-parameters-for-the-functions-in-the-package">
<h3>Detailed information about the parameters for the functions in the package<a class="headerlink" href="#detailed-information-about-the-parameters-for-the-functions-in-the-package" title="Permalink to this heading">¶</a></h3>
</section>
<section id="get-singlets">
<h3>get_singlets<a class="headerlink" href="#get-singlets" title="Permalink to this heading">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="source.singletCode.singletCode.get_singlets">
<span class="sig-name descname"><span class="pre">get_singlets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sheet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_all_singlet_categories</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_plot_umi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_cutoff_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ratio'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_cutoff_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7.5e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_cutoff_percentile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_umi_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_diff_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_dominant_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#source.singletCode.singletCode.get_singlets" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that inputs the sample sheet and other parameters and runs it through count_doublets to get a list of
singlets in the sample. If a row is repeated, it is assumed to reflect the UMI associated with a barcode in
this cell (identified by the cellID).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sheet</strong> (<em>DataFrame</em>) – A dataframe that contains 3 columns: cellID, barcode, sample.</p></li>
<li><p><strong>dataset_name</strong> (<em>str</em>) – The name of the dataset being analysed. It will be in the name of all saved files
and be a column in the singlet_stats sheet returned.</p></li>
<li><p><strong>output_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to store any output files, including plots to show UMI distribution
and what the umi_cutoff used is, csv files containing singlets of different categories. If None,
then the list of singlets will be returned but it won’t contain information about what category of
singlet each cell is. Defaults to None.</p></li>
<li><p><strong>save_all_singlet_categories</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, then singlets of each category are saved separately
in csv files along with all singlets and all non-singlets. Defaults to False.</p></li>
<li><p><strong>save_plot_umi</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, then plots showing UMI distribution indicating the UMI cutoff used
will be saved for each sample. Defaults to False.</p></li>
<li><p><strong>umi_cutoff_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Specify if quality control for barcodes using UMI counts should be based
on “ratio” or “percentile”. Defaults to ‘ratio’.</p></li>
<li><p><strong>umi_cutoff_ratio</strong> (<em>float</em><em>, </em><em>optional</em>) – The ratio used to determine the umi_cutoff if umi_cutoff_method is
“ratio”. Defaults to 3/4e5.</p></li>
<li><p><strong>umi_cutoff_percentile</strong> (<em>float</em><em>, </em><em>optional</em>) – If umi_cutoff_method is “percentile”, then the umi_cutoff will be
the minimum UMI count required to be in the top umi_cutoff_percentile’th percentile. There is no default and if umi_cutoff_method is set to “percentile”, then manually set this parameter.</p></li>
<li><p><strong>min_umi_cutoff</strong> (<em>int</em><em>, </em><em>optional</em>) – This is the absolute minimum number of UMIs that need to be associated with
a barcode for it to be considered a barcode. However, the actual umi_cutoff used will be the greater
of min_umi_cutoff and the cutoff calculated using umi_cutoff_method. Defaults to 2.</p></li>
<li><p><strong>umi_dominant_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum UMI count to be associated with a barcode for it to be
considered to be a potential dominant barcode in a cell. Defaults to 10.</p></li>
<li><p><strong>umi_diff_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – This is the minimum difference between UMI counts associated with a
potential dominant barcode present within a cell and the median UMI count of all barcodes associated
with the cell. If a cell has only one dominant barcode, it will be counted. Defaults to 50.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>A 2-tuple containing:</dt><dd><ul class="simple">
<li><p>pandas.DataFrame: A dataframe which contains all unique cell ID and barcode combinations in the data along with singlet assignment to each cell ID.</p></li>
<li><p>pandas.DataFrame: A dataframe which contains the statistics for total singlets, different categories of singlets, and cells removed due to low UMI counts.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="check-sample-sheet">
<h3>check_sample_sheet<a class="headerlink" href="#check-sample-sheet" title="Permalink to this heading">¶</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="source.singletCode.singletCode.check_sample_sheet">
<span class="sig-name descname"><span class="pre">check_sample_sheet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_frame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#source.singletCode.singletCode.check_sample_sheet" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to check if the dataframe can be used as input to get_singlets function. It checks if the three columns - cellID, barcode, sample needed are present and in same order. If this dataframe can be used for get_singlets function, then a statement will be printed to confirm that</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_sheet</strong> – A dataframe that contains your sample sheet.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
<span id="document-infoCLI"></span><section id="command-line-interface">
<h2>command line interface<a class="headerlink" href="#command-line-interface" title="Permalink to this heading">¶</a></h2>
<section id="availablility">
<h3>Availablility<a class="headerlink" href="#availablility" title="Permalink to this heading">¶</a></h3>
<p>The source code for the command line tool is available in our GitHub repository (<a class="reference external" href="https://github.com/GoyalLab/singletCodeTools/">repo</a>). You can clone the reposiory to access the code locally.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/GoyalLab/singletCodeTools/
</pre></div>
</div>
</section>
<section id="using-the-interface">
<h3>Using the interface<a class="headerlink" href="#using-the-interface" title="Permalink to this heading">¶</a></h3>
<p>Navigating to commandLineTools folder, you will find 3 files and you will need to run  singletCodeCommandLine.py. There are 2 modules available, one to run singletCode and the other to create a sample sheet if you have used Watermelon barcoding techmology using the fastq sequenced files from MISEQ.</p>
</section>
<section id="module-source.commandLineTools.singletCodeCommandLine">
<span id="detailed-information-about-the-modules-in-the-command-line-interface"></span><h3>Detailed information about the modules in the command line interface<a class="headerlink" href="#module-source.commandLineTools.singletCodeCommandLine" title="Permalink to this heading">¶</a></h3>
<dl>
<dt>Description:</dt><dd><p>This script contains two modules:
- Count Module: Generates the singlet files for the input data sheet.
- Watermelon Module: Uses the MiSeq dial-out files to create the cell ID, barcode, and sample file. These outputs can then be used as input for the singlet code module.</p>
</dd>
<dt>Usage:</dt><dd><p>For the Count Module:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>singletCode.py<span class="w"> </span>count<span class="w"> </span>-i<span class="w"> </span>/path/to/input.txt<span class="w"> </span>-o<span class="w"> </span>/path/to/output
</pre></div>
</div>
<p>For the Watermelon Module:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>singletCodeCommadLine.py<span class="w"> </span>watermelon<span class="w"> </span>-i<span class="w"> </span>/path/to/fastq/files<span class="w"> </span>-o<span class="w"> </span>path/to/save/csv/file<span class="w"> </span>-s<span class="w"> </span>path/sample/sheet<span class="w"> </span>-use10X<span class="w"> </span>False<span class="w"> </span>-input10X<span class="w"> </span>path/to/barcodes/tsv
</pre></div>
</div>
</dd>
<dt>Options for Count Module:</dt><dd><ul class="simple">
<li><p>-i, –input_file: Specify the path to the input barcode file.</p></li>
<li><p>-o, –output_prefix: Specify the path to the output prefix.</p></li>
<li><p>-f, –force: Force overwrite if the output file already exists.</p></li>
<li><p>-u, –cutoff: UMI cutoff ratio.</p></li>
<li><p>-d, –umi_diff_threshold: Minimum difference in UMI between UMI count for dominant UMI and median UMI count.</p></li>
<li><p>-m, –min_umi_good_data_cutoff: Minimum UMI count for a barcode to be dominant.</p></li>
<li><p>-g, –min_umi_filter_threshold: Minimum UMI filter threshold.</p></li>
</ul>
</dd>
<dt>Options for Watermelon Module:</dt><dd><ul class="simple">
<li><p>-i, –inputFolder: Specify the path to the folder containing the fastq folders output from MiSeq.</p></li>
<li><p>-s, –sampleSheet: Specify the path to the sample sheet in .csv format that contains sample name and sample number. This should match the names of the fastq files (e.g., Sample-1_S1_L001_R1_001.fastq.gz).</p></li>
<li><p>-o, –outputFolder: Specify the path to save the output CSV file containing the barcode, cell ID information.</p></li>
<li><p>-outputName: Specify the name of the output CSV file.</p></li>
<li><p>–use10X: Specify if a 10X object is provided which has the same cells as those in fastq. If provided, cells in the fastq file will be filtered out if not present in the 10X object.</p></li>
<li><p>–input10X: Path to the barcodes.tsv.gz or barcodes.tsv which contains cell IDs.</p></li>
</ul>
</dd>
<dt>Authors:</dt><dd><ul class="simple">
<li><p>Ziyang Zhang (Charles)</p></li>
<li><p>Keerthana M Arun</p></li>
</ul>
</dd>
</dl>
</section>
</section>
<span id="document-watermelonDatasetVignette"></span><section id="vignette-to-use-singletcode-command-line-tool-to-analyse-single-cell-data-with-watermelon-barcodes">
<h2>Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes<a class="headerlink" href="#vignette-to-use-singletcode-command-line-tool-to-analyse-single-cell-data-with-watermelon-barcodes" title="Permalink to this heading">¶</a></h2>
<p>The barcode region is assumed to be amplified using Illumina MiSeq.</p>
<p>All the data for this vignette and the files output from it can be
downloaded
<a class="reference external" href="https://github.com/GoyalLab/SingletCodeWebsite/raw/main/source/dataVignette/watermelonVignetteData.zip">here</a>.
It contains inputFiles and the outputFiles (it contains a test folder
which has the expected output files). This vignette can be downloaded as
a jupyter notebook from the <a class="reference external" href="https://github.com/GoyalLab/singletCodeTools/tree/main/vignette">singletCode Tools
repo</a>.</p>
<section id="installing-singletcode-command-line-tool">
<h3>Installing singletCode Command line tool<a class="headerlink" href="#installing-singletcode-command-line-tool" title="Permalink to this heading">¶</a></h3>
<p>To use the singletCode command line tool, clone the repository from
GitHub. Let the Path to the folder you are running this command be
<strong>Path</strong>. We can also install other packages needed to run the tool.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/GoyalLab/singletCodeTools
<span class="n">Path</span> <span class="o">=</span> <span class="s2">&quot;path/to/singletCodeTools/repo&quot;</span>
<span class="o">%</span><span class="k">conda</span> install scipy tqdm matplotlib biopython python-levenshtein pandas
</pre></div>
</div>
</section>
<section id="next-step-is-to-understand-the-samples-present-in-the-fastq-files">
<h3>Next step is to understand the samples present in the FASTQ files.<a class="headerlink" href="#next-step-is-to-understand-the-samples-present-in-the-fastq-files" title="Permalink to this heading">¶</a></h3>
<p>The sample fastq files are in the inputFolder. We can identify the
sample name and number from the FASTQ file. For example,
sampleName_S1_L001_R1_001.fastq.gz means that the sample name is
sampleName and sample number is 1.Make sure that both read 1 and read 2
for each sample are present in the same folder (R1 and R2)</p>
</section>
<section id="creating-sample-sheet-for-these-two-samples">
<h3>Creating sample sheet for these two samples.<a class="headerlink" href="#creating-sample-sheet-for-these-two-samples" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="s2">&quot;path/to/downloaded/and/unzipped/data&quot;</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sampleSheet</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/inputFiles/sampleSheet.csv&quot;</span><span class="p">)</span>
<span class="n">sampleSheet</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sampleName</th>
      <th>sampleNumber</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>sampleName</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>otherSampleName</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div><p>Now, to run the watermelon module of singletCodeTools, you need to run
this command. If we are going by the folder structure of the zipped file
and <strong>p</strong> is <em>path to the unzipped folder containing example files</em>,
then 1. <strong>inputFolder</strong> will be p/inputFiles/ 2. <strong>outputFolder</strong> will
be p/outputFiles/ 3. <strong>sampleSheet</strong> will be
p/inputFiles/sampleSheet.csv</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">subprocess</span>

<span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s1">&#39;python&#39;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">Path</span><span class="si">}</span><span class="s1">/commandLine/singletCodeCommandLine.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;watermelon&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-i&#39;</span><span class="p">,</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/sampleSheet.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--outputName&#39;</span><span class="p">,</span> <span class="s1">&#39;watermelonBarcodeUmi.csv&#39;</span>
<span class="p">]</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Arguments</span> <span class="n">received</span><span class="p">:</span>
  <span class="n">command</span><span class="p">:</span> <span class="n">watermelon</span>
  <span class="n">inputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span>
  <span class="n">outputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">outputFiles</span>
  <span class="n">sampleSheet</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span><span class="n">sampleSheet</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">outputName</span><span class="p">:</span> <span class="n">watermelonBarcodeUmi</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">use10X</span><span class="p">:</span> <span class="kc">False</span>
  <span class="n">input10X</span><span class="p">:</span> <span class="kc">None</span>
<span class="n">All</span> <span class="n">the</span> <span class="n">inputs</span> <span class="k">for</span> <span class="n">the</span> <span class="n">command</span> <span class="n">are</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">will</span> <span class="n">proceed</span> <span class="k">with</span> <span class="n">creating</span> <span class="n">the</span> <span class="n">barcode</span> <span class="n">sheet</span> <span class="k">for</span> <span class="nb">all</span> <span class="n">the</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">sheet</span><span class="o">.</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">940</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">718</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
    <span class="s1">&#39;python&#39;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">Path</span><span class="si">}</span><span class="s1">/commandLine/singletCodeCommandLine.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;watermelon&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/sampleSheet.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--outputName&#39;</span><span class="p">,</span> <span class="s1">&#39;watermelonBarcodeUmiWith10X.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--use10X&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--input10X&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/barcodes.tsv&#39;</span>
<span class="p">],</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Check if the command was successful</span>
<span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">returncode</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command executed successfully&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Output:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command failed&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Command</span> <span class="n">executed</span> <span class="n">successfully</span>
<span class="n">Output</span><span class="p">:</span>
 <span class="n">Arguments</span> <span class="n">received</span><span class="p">:</span>
  <span class="n">command</span><span class="p">:</span> <span class="n">watermelon</span>
  <span class="n">inputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span>
  <span class="n">outputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">outputFiles</span><span class="o">/</span>
  <span class="n">sampleSheet</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span><span class="n">sampleSheet</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">outputName</span><span class="p">:</span> <span class="n">watermelonBarcodeUmiWith10X</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">use10X</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">input10X</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span><span class="n">barcodes</span><span class="o">.</span><span class="n">tsv</span>
<span class="n">All</span> <span class="n">the</span> <span class="n">inputs</span> <span class="k">for</span> <span class="n">the</span> <span class="n">command</span> <span class="n">are</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">will</span> <span class="n">proceed</span> <span class="k">with</span> <span class="n">creating</span> <span class="n">the</span> <span class="n">barcode</span> <span class="n">sheet</span> <span class="k">for</span> <span class="nb">all</span> <span class="n">the</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">sheet</span><span class="o">.</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">791</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">629</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">subprocess</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
    <span class="s1">&#39;python&#39;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">Path</span><span class="si">}</span><span class="s1">/commandLine/singletCodeCommandLine.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;count&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/watermelonBarcodeUmiWith10X.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/watermelon&#39;</span>
<span class="p">],</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Check if the command was successful</span>
<span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">returncode</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command executed successfully&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Output:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command failed&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Command executed successfully
Output:
 Arguments received:
  command: count
  input_file: /home/keerthana/Goyal_Lab/websiteToolData/thingsToAddToWebsite/watermelonVignetteData/watermelonVignetteData/outputFiles/watermelonBarcodeUmiWith10X.csv
  out_prefix: /home/keerthana/Goyal_Lab/websiteToolData/thingsToAddToWebsite/watermelonVignetteData/watermelonVignetteData/outputFiles/watermelon
  umi_cutoff_ratio: 7.5e-06
  umi_diff_threshold: 50
  dominant_threshold: 10
  min_umi_good_data_cutoff: 2
INFO: Raw data counts
sampleNum
sampleName         693
otherSampleName    524
Name: count, dtype: int64
INFO: Using raio based filtering.
Current Sample Adjusted UMI cutoff: 2
Total cells: 45
Sample sampleName singlet: 43
Total Singlets: 43
Total Multiplets: 1
All singlets identified are unique? True
Total Singlets: 43
Total Multiplets: 1
INFO: Using raio based filtering.
Current Sample Adjusted UMI cutoff: 2
Total cells: 22
Sample otherSampleName singlet: 22
Total Singlets: 22
Total Multiplets: 0
All singlets identified are unique? True
Total Singlets: 22
Total Multiplets: 0
All singlets identified are unique? True
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_sampleName_singlets_stats.csv&quot;</span><span class="p">)</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#62575b&#39;</span><span class="p">,</span> <span class="s1">&#39;#2175a8&#39;</span><span class="p">,</span> <span class="s1">&#39;#feb422&#39;</span><span class="p">,</span> <span class="s1">&#39;#d62728&#39;</span><span class="p">,</span> <span class="s1">&#39;#d4d4d4&#39;</span><span class="p">]</span>  <span class="c1"># Example colors, modify as needed</span>
<span class="n">plotData</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;total_cells&#39;</span><span class="p">,</span> <span class="s2">&quot;total_singlets&quot;</span><span class="p">])</span>

<span class="c1"># Plotting</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plotData</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;barh&#39;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>

<span class="k">for</span> <span class="n">plot</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">patches</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">plot</span><span class="o">.</span><span class="n">get_width</span><span class="p">())</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="p">(</span><span class="n">plot</span><span class="o">.</span><span class="n">get_x</span><span class="p">()</span> <span class="o">+</span> <span class="n">plot</span><span class="o">.</span><span class="n">get_width</span><span class="p">()</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">plot</span><span class="o">.</span><span class="n">get_y</span><span class="p">()</span> <span class="o">+</span> <span class="n">plot</span><span class="o">.</span><span class="n">get_height</span><span class="p">()</span><span class="o">/</span><span class="mi">2</span><span class="p">),</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Total cells&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Singlets by Criteria&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<img alt="_images/watermelonDatasetVignette_10_0.png" src="_images/watermelonDatasetVignette_10_0.png" />
<p>In the above plot, you see that the original data had 569 cells that
were removed due to low barcode UMI count, 43 singlets with a
single-barcode associated with them and 1 multiplet (singletCode could
not determine if it was a singlet for sure.)</p>
</section>
<section id="looking-at-the-scrnaseq-data-associated">
<h3>Looking at the scRNAseq data associated<a class="headerlink" href="#looking-at-the-scrnaseq-data-associated" title="Permalink to this heading">¶</a></h3>
<p>Since this data has both scRNAseq and barcodes for the same cells, we
can analyse them together</p>
<p>Installing and importing scanpy package to do this</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Install scanpy for further single-cell RNAseq analysis</span>
<span class="c1"># %conda install -c conda-forge scanpy python-igraph leidenalg</span>
<span class="c1">#Import scanpy</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
</pre></div>
</div>
<p>In case there are version conflicts during this installation or while
importing scanpy, we found <em>%conda update –all</em> to be an useful command
that fixed the version conflict previously. Reading in the 10X h5ad
object associated with the same watermelon data</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/inputFiles/watermelonScRnaSeqData.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1093 × 27264
</pre></div>
</div>
</section>
<section id="read-in-the-output-files-to-identify-cells-as-being-singlets-multiplets-or-being-removed-for-low-barcode-umi-threshold">
<h3>Read in the output files to identify cells as being singlets, multiplets or being removed for low barcode UMI threshold<a class="headerlink" href="#read-in-the-output-files-to-identify-cells-as-being-singlets-multiplets-or-being-removed-for-low-barcode-umi-threshold" title="Permalink to this heading">¶</a></h3>
<p>First, reading in the cellID-barcode-UMI sheet generated earlier with
additional filter using scRNAseq data</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellidBarcodeUMI</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/watermelonBarcodeUmiWith10X.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Reading in all the singlets and multiplets idenified in the two samples.
There might not always be multiplets - check the stats file to see if
there are any. In this example, there are no multiplets in
otherSampleName.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sampleNameSinglets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_sampleName_singlets_all.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">otherSampleNameSinglets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_otherSampleName_singlets_all.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">sampleNameMultiplets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_sampleName_multiplets.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<p>Identifying the cells that were below the barcode UMI threshold and were
filtered out by singletCode</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lowUmiCells</span> <span class="o">=</span> <span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">|</span>
                                 <span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">otherSampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">|</span>
                                 <span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameMultiplets</span><span class="p">[</span><span class="mi">0</span><span class="p">]))]</span>
</pre></div>
</div>
</section>
<section id="annotating-the-cells-in-adata-with-these-labels">
<h3>Annotating the cells in adata with these labels<a class="headerlink" href="#annotating-the-cells-in-adata-with-these-labels" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Annotating the adata with these labels using the lists created</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;singlet&#39;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">otherSampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;singlet&#39;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameMultiplets</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;multiplet&#39;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">lowUmiCells</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;low UMI&#39;</span>
</pre></div>
</div>
<p><strong>Note</strong> that in this vignette we are not doing any actual QC - but in
actual analysis, it would need to be done.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-pca-and-umap-for-visualization">
<h3>Calculating PCA and UMAP for visualization<a class="headerlink" href="#calculating-pca-and-umap-for-visualization" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculating PCA for the data and plotting variance ratio</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/watermelonDatasetVignette_26_0.png" src="_images/watermelonDatasetVignette_26_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">],</span>
    <span class="n">size</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<img alt="_images/watermelonDatasetVignette_27_0.png" src="_images/watermelonDatasetVignette_27_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculating neighbours and UMAP from that for further visualization</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">random_state</span> <span class="o">=</span> <span class="mi">101010</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;singletStatus&#39;</span><span class="p">],</span>
    <span class="n">size</span><span class="o">=</span><span class="mi">60</span>
<span class="p">)</span>
</pre></div>
</div>
<img alt="_images/watermelonDatasetVignette_29_0.png" src="_images/watermelonDatasetVignette_29_0.png" />
</section>
<section id="saving-the-final-adata">
<h3>Saving the final adata<a class="headerlink" href="#saving-the-final-adata" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelonScRNA.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<span id="document-singletCodePackageVignette"></span><section id="vignette-to-use-singletcode-package">
<h2>Vignette to use singletCode package<a class="headerlink" href="#vignette-to-use-singletcode-package" title="Permalink to this heading">¶</a></h2>
<p>The input needed to run singletCode is a .csv file that contains the
information about cell ID (added while sequencing), lineage barcode, and
sample name. Each row should be repeated n times where n is the number
of UMIs associated with that barcode and cell ID combination. You can
download a sample input sheet
<a class="reference external" href="https://github.com/GoyalLab/SingletCodeWebsite/raw/main/source/dataVignette/singletCodePackageVignetteData.zip">here</a>.
It is a subset of data from Jiang Et al and details about it are
described in the singletCode paper in detail. The folder also contains
expected output files in the test folder within the outputFiles folder.
This vignette can be downloaded as a jupyter notebook from the
<a class="reference external" href="https://github.com/GoyalLab/singletCodeTools/tree/main/vignette">singletCode Tools
repo</a>.</p>
<section id="install-singletcode-package">
<h3>Install singletCode package<a class="headerlink" href="#install-singletcode-package" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip3<span class="w"> </span>install<span class="w"> </span>singletCode
</pre></div>
</div>
<p>Import necessary functions from it</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">singletCode</span> <span class="kn">import</span> <span class="n">check_sample_sheet</span><span class="p">,</span> <span class="n">get_singlets</span>
</pre></div>
</div>
</section>
<section id="read-in-input-sheet">
<h3>Read in input sheet<a class="headerlink" href="#read-in-input-sheet" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in input sheet</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;path/to/downloaded/and/unzipped/folder&quot;</span>
<span class="n">pathToInputSheet</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">/inputFiles/JiangEtAlSubset_InputSheet.csv&quot;</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pathToInputSheet</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="check-formatting-of-input-sheet">
<h3>Check formatting of input sheet<a class="headerlink" href="#check-formatting-of-input-sheet" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">check_sample_sheet</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">The</span> <span class="n">sample</span> <span class="n">sheet</span> <span class="n">provided</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="k">as</span> <span class="nb">input</span> <span class="n">to</span> <span class="n">get_singlets</span> <span class="n">to</span> <span class="n">get</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">singlets</span> <span class="n">identified</span><span class="o">.</span>
</pre></div>
</div>
<p>The sample sheet provided can be used as input to get_singlets to get a
list of singlets identified.</p>
</section>
<section id="identify-singlets-from-input-sheet">
<h3>Identify singlets from input sheet<a class="headerlink" href="#identify-singlets-from-input-sheet" title="Permalink to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outputPath</span> <span class="o">=</span> <span class="s2">&quot;path/to/output/folder&quot;</span>
<span class="n">cellLabelList</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="n">get_singlets</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">dataset_name</span><span class="o">=</span> <span class="s2">&quot;JiangEtAlSubset&quot;</span><span class="p">,</span> <span class="n">save_all_singlet_categories</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="n">outputPath</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span><span class="p">:</span> <span class="n">Raw</span> <span class="n">data</span> <span class="n">counts</span><span class="p">:</span>
<span class="n">sample</span>
<span class="mi">1</span>    <span class="mi">1306</span>
<span class="n">Name</span><span class="p">:</span> <span class="n">count</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
<span class="n">Total</span> <span class="n">cells</span> <span class="k">for</span> <span class="n">sample</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">39</span>
<span class="n">INFO</span><span class="p">:</span> <span class="n">Using</span> <span class="n">ratio</span> <span class="n">based</span> <span class="n">filtering</span><span class="o">.</span>
<span class="n">Current</span> <span class="n">Sample</span> <span class="n">Adjusted</span> <span class="n">UMI</span> <span class="n">cutoff</span><span class="p">:</span> <span class="mi">2</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>100%|██████████| 122/122 [00:00&lt;00:00, 11475.27it/s]
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>All singlets identified with multiple barcodes are unique? True
Total Singlets: 10
Total Multiplets: 9
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">stats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outputPath</span><span class="si">}</span><span class="s2">/JiangEtAlSubset_stats.csv&quot;</span><span class="p">)</span>
<span class="n">cellLabelList</span><span class="p">[</span><span class="n">cellLabelList</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Singlet&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outputPath</span><span class="si">}</span><span class="s2">/JiangEtAlSubset_singletList.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Visualizing the distribution of cells into low UMI, different kinds of
singlets and undetermined</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="c1">#Plotting the distribution of low UMI cells, different kinds of singlets, and undetermined cells.</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#62575b&#39;</span><span class="p">,</span> <span class="s1">&#39;#2175a8&#39;</span><span class="p">,</span> <span class="s1">&#39;#feb422&#39;</span><span class="p">,</span> <span class="s1">&#39;#d62728&#39;</span><span class="p">,</span> <span class="s1">&#39;#d4d4d4&#39;</span><span class="p">]</span>  <span class="c1"># Example colors, modify as needed</span>
<span class="n">plotData</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;total_cells&#39;</span><span class="p">,</span> <span class="s2">&quot;total_singlets&quot;</span><span class="p">])</span>

<span class="c1"># Plotting</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plotData</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;barh&#39;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>

<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">patches</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">get_width</span><span class="p">())</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">get_x</span><span class="p">()</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">get_width</span><span class="p">()</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">get_y</span><span class="p">()</span> <span class="o">+</span> <span class="n">p</span><span class="o">.</span><span class="n">get_height</span><span class="p">()</span><span class="o">/</span><span class="mi">2</span><span class="p">),</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Total cells&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Singlets by Criteria&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<img alt="_images/singletCodePackageVignette_12_0.png" src="_images/singletCodePackageVignette_12_0.png" />
<p>The above plot shows that the data we had contained different kind of
singlets: 6 single-barcode cells, 2 cells which had more than one
barcode but with same combination being present in more than one cell, 2
cells that had one dominant barcode. The data also contained 9 cells
which singletCode could not determine as being truly singlets and 20
cells whose barcode UMI counts were below the set threshold.</p>
</section>
<section id="understanding-the-output-files">
<h3>Understanding the output files<a class="headerlink" href="#understanding-the-output-files" title="Permalink to this heading">¶</a></h3>
<p>To understand some of the files in the output, we can look at cell IDs
and their data in the original input sheet</p>
<p>For the dominant_umi_singlets, there are two cell IDs. One of them is
TGTAAGCGTCTCGCGA. If we look at that entry in the input sheet and count
the number of UMI associated with each barcode, we see that one barcode
has 99 UMI counts while the second highest UMI count is 7. So, the cell
most likely has only one barcode associated with it and hence, a
singlet.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TGTAAGCGTCTCGCGA&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cellID&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cellID</th>
      <th>barcode</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATTGTTGTTGCAGATGCAGTTGATGCTGATGAAGTTGTACAAGGTC...</td>
      <td>1</td>
      <td>99</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATTCGACTTGATCTTCTAGAACATGGTGAACTAGCAGGTGCTGATC...</td>
      <td>1</td>
      <td>7</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATACTAGCTCAAGCAGTACTACTACTTCGTCTTCATGCAGAACAAC...</td>
      <td>1</td>
      <td>6</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATAGATGCACTTGGTGGTCGAGTTCTAGTTGTAGCTGATCGTCCAG...</td>
      <td>1</td>
      <td>6</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATTCGACCAGAACCACATGCAGTTCAACGTGTTCGAGGTGTAGATG...</td>
      <td>1</td>
      <td>6</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>82</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATAGTAGTAGCTGTTGGTGTTGAAGTACTTCCTCTTGCTCCTCGTG...</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>83</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATAGTAGATGAACGTCCTCTACATGTTCTTCGTCAAGTACCAGCAC...</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>84</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATAGTACATGGTGGACCTGGACTTCGAGATGGAGCTCTTGTTCCTG...</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>85</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>ATAGGAGTAGTTGGTGATGGTCTACCAGAAGGTGAAGGTGGAGAAG...</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>86</th>
      <td>TGTAAGCGTCTCGCGA</td>
      <td>GGTGCTCAACTTCTTGTTGTACTTCTAGTTGATGTTGGACGTCATC...</td>
      <td>1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>87 rows × 4 columns</p>
</div><p>Next, we can look at multi-barcode singlets. There are two cell IDs:
AGGCTGCTCTTTCCGG and GAGGGATGTAACATCC. If we look at the barcodes with
greater than 2 UMI counts, we see that they have the same combination.
The only way this can occur is if a cell receives multiple barcode
initially and then divides.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;AGGCTGCTCTTTCCGG&#39;</span><span class="p">]</span>
 <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;cellID&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
 <span class="o">.</span><span class="n">size</span><span class="p">()</span>
 <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">)</span>
 <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
 <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;count &gt;= 2&#39;</span><span class="p">)</span>
 <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cellID</th>
      <th>barcode</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AGGCTGCTCTTTCCGG</td>
      <td>ATAGGAGTAGTTGGTGATGGTCTACCAGAAGGTGAAGGTGGAGAAG...</td>
      <td>1</td>
      <td>13</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AGGCTGCTCTTTCCGG</td>
      <td>ATTGAACGTGGAGTTGAACTTGTACTACGAGTACGTCTAGAACATG...</td>
      <td>1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></section>
<section id="scrnaseq-data">
<h3>scRNAseq data<a class="headerlink" href="#scrnaseq-data" title="Permalink to this heading">¶</a></h3>
<p>Further single-cell RNAseq analysis with both scRNAseq data and singlet
information from singletCode output</p>
<p>Install and import scanpy for further single-cell RNAseq analysis</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>scanpy<span class="o">[</span>leiden<span class="o">]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Import scanpy</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
</pre></div>
</div>
<p>Reading the scRNAseq input data in h5ad format</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Reading the scRNAseq data in h5ad format</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">/inputFiles/JiangEtAlSubset_scRNAseqData.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 39 × 36601
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;
</pre></div>
</div>
<div class="line-block">
<div class="line">Making copies of singletCode input/output to use them along with
scRNAseq data. The -1 is added to cell IDs to match the cell IDs seen
in 10x format data.</div>
<div class="line"><strong>NOTE</strong>: It may not be needed for your actual data.</div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">singleCellDf</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">singleCellDf</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">singleCellDf</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;-1&quot;</span>
<span class="n">singleCellDf</span> <span class="o">=</span> <span class="n">singleCellDf</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;cellID&#39;</span><span class="p">)</span>
<span class="n">cellLabelListSingleCell</span> <span class="o">=</span> <span class="n">cellLabelList</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">cellLabelListSingleCell</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cellLabelListSingleCell</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;-1&quot;</span>
<span class="n">cellLabelListSingleCell</span> <span class="o">=</span> <span class="n">cellLabelListSingleCell</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;cellID&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<section id="calculating-total-counts-and-genes-identified-per-cell">
<h4>Calculating total counts and genes identified per cell.<a class="headerlink" href="#calculating-total-counts-and-genes-identified-per-cell" title="Permalink to this heading">¶</a></h4>
<p><strong>NOTE</strong>: In this vignette we are not doing any actual QC - but in
actual analysis, it would need to be done.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-pca-and-plotting-variance-ratio-vs-ranking">
<h4>Calculating PCA and plotting variance ratio vs ranking<a class="headerlink" href="#calculating-pca-and-plotting-variance-ratio-vs-ranking" title="Permalink to this heading">¶</a></h4>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/singletCodePackageVignette_28_0.png" src="_images/singletCodePackageVignette_28_0.png" />
<p>Identifying cells that were thresholded by singletCode as low UMI by
identifying cells that were in the original list provided to singletCode
but not labeled as either singlet or undetermined. Then creating a list
of annotations of singletStatus(singlet, multiplet, low UMI) for all
cells</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">umiCutoff</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">singleCellDf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">singleCellDf</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cellLabelListSingleCell</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]),</span> <span class="s1">&#39;cellID&#39;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">umiCutoff</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Low UMI&quot;</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellIDLabels</span> <span class="o">=</span> <span class="n">cellLabelListSingleCell</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;nUMI&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Creating a list of cell IDs with annotation of whether singlet, multiplet or low UMI.</span>
<span class="n">labelID</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">umiCutoff</span><span class="p">,</span> <span class="n">cellIDLabels</span><span class="p">])</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">labelID</span> <span class="o">=</span> <span class="n">labelID</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">labelID</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">])</span>
<span class="c1">#Adding the labels to cells in the adata to visualise it</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;singletStatus&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">labelID</span>
</pre></div>
</div>
</section>
<section id="visualising-the-cells-in-pca-space">
<h4>Visualising the cells in PCA space<a class="headerlink" href="#visualising-the-cells-in-pca-space" title="Permalink to this heading">¶</a></h4>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">],</span>
    <span class="n">size</span> <span class="o">=</span> <span class="mi">250</span>
<span class="p">)</span>
</pre></div>
</div>
<img alt="_images/singletCodePackageVignette_34_0.png" src="_images/singletCodePackageVignette_34_0.png" />
</section>
<section id="calculating-neigbours-and-umap-for-further-visualisation">
<h4>Calculating neigbours and UMAP for further visualisation<a class="headerlink" href="#calculating-neigbours-and-umap-for-further-visualisation" title="Permalink to this heading">¶</a></h4>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">101010</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;singletStatus&#39;</span><span class="p">],</span>
    <span class="c1"># Setting a smaller point size to get prevent overlap</span>
    <span class="n">size</span><span class="o">=</span><span class="mi">250</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<img alt="_images/singletCodePackageVignette_37_0.png" src="_images/singletCodePackageVignette_37_0.png" />
</section>
<section id="saving-the-anndata">
<h4>Saving the AnnData<a class="headerlink" href="#saving-the-anndata" title="Permalink to this heading">¶</a></h4>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">outputPath</span><span class="si">}</span><span class="s2">/JiangEtAlSubset.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>
<span id="document-Contact"></span><section id="questions-contact-us">
<h2>questions? contact us!<a class="headerlink" href="#questions-contact-us" title="Permalink to this heading">¶</a></h2>
<p>Do you have a doublet detection method that you would like use with singletCode? Do you have a question about how to use singletCode? Do you have a suggestion for how to improve singletCode? Please contact us!</p>
<p>For more information or questions, please contact Yogesh Goyal at <a class="reference external" href="mailto:yogesh&#46;goyal&#37;&#52;&#48;northwestern&#46;edu">yogesh<span>&#46;</span>goyal<span>&#64;</span>northwestern<span>&#46;</span>edu</a></p>
</section>
<span id="document-repositories"></span><section id="repositories">
<h2>repositories<a class="headerlink" href="#repositories" title="Permalink to this heading">¶</a></h2>
<p>GitHub repository for analyses performed in our paper <a class="reference external" href="https://github.com/GoyalLab/fatemap_multiplet_public">here</a>.</p>
<p>GitHub repository for singletCode tools (command line interface and python package) <a class="reference external" href="https://github.com/GoyalLab/singletCodeTools">here</a>.</p>
</section>
</div>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Goyal Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>