<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes &mdash; singletCode 1.2 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/sphinx_rtd_size.css" type="text/css" />
      <link rel="stylesheet" href="../_static/colour.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script type="text/javascript">document.addEventListener('DOMContentLoaded', function() {document.querySelectorAll('a.reference.external').forEach(function(a) {a.setAttribute('target', '_blank'); a.setAttribute('rel', 'noopener noreferrer');});});</script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex/" />
    <link rel="search" title="Search" href="../search/" />
    <link rel="next" title="Vignette to use singletCode package" href="../singletCodePackageVignette/" />
    <link rel="prev" title="command line interface" href="../infoCLI/" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../" class="icon icon-home">
            singletCode
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search/" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../doubletInformation/">doublets in sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../singletCode/">what is singletCode?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../classifier/">classifier</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usageInformation/">implementing singletCode on your data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../doubletDetectionMethods/">benchmarking doublet detection methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../datasetsInfo/">dataset information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../packageDocumentation/">singletCode PyPI package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../infoCLI/">command line interface</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installing-singletcode-command-line-tool">Installing singletCode Command line tool</a></li>
<li class="toctree-l2"><a class="reference internal" href="#next-step-is-to-understand-the-samples-present-in-the-fastq-files">Next step is to understand the samples present in the FASTQ files.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-sample-sheet-for-these-two-samples">Creating sample sheet for these two samples.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#looking-at-the-scrnaseq-data-associated">Looking at the scRNAseq data associated</a></li>
<li class="toctree-l2"><a class="reference internal" href="#read-in-the-output-files-to-identify-cells-as-being-singlets-multiplets-or-being-removed-for-low-barcode-umi-threshold">Read in the output files to identify cells as being singlets, multiplets or being removed for low barcode UMI threshold</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotating-the-cells-in-adata-with-these-labels">Annotating the cells in adata with these labels</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-pca-and-umap-for-visualization">Calculating PCA and UMAP for visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#saving-the-final-adata">Saving the final adata</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../singletCodePackageVignette/">Vignette to use singletCode package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Contact/">questions? contact us!</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.cell.com/cell-genomics/fulltext/S2666-979X(24)00176-9">link to paper</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.goyallab.org/">Goyal Lab website</a></li>
<li class="toctree-l1"><a class="reference internal" href="../repositories/">repositories</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../">singletCode</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/watermelonDatasetVignette.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="vignette-to-use-singletcode-command-line-tool-to-analyse-single-cell-data-with-watermelon-barcodes">
<h1>Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes<a class="headerlink" href="#vignette-to-use-singletcode-command-line-tool-to-analyse-single-cell-data-with-watermelon-barcodes" title="Permalink to this heading">¶</a></h1>
<p>The barcode region is assumed to be amplified using Illumina MiSeq.</p>
<p>All the data for this vignette and the files output from it can be
downloaded
<a class="reference external" href="https://github.com/GoyalLab/SingletCodeWebsite/raw/main/source/dataVignette/watermelonVignetteData.zip">here</a>.
It contains inputFiles and the outputFiles (it contains a test folder
which has the expected output files). This vignette can be downloaded as
a jupyter notebook from the <a class="reference external" href="https://github.com/GoyalLab/singletCodeTools/tree/main/vignette">singletCode Tools
repo</a>.</p>
<section id="installing-singletcode-command-line-tool">
<h2>Installing singletCode Command line tool<a class="headerlink" href="#installing-singletcode-command-line-tool" title="Permalink to this heading">¶</a></h2>
<p>To use the singletCode command line tool, clone the repository from
GitHub. Let the Path to the folder you are running this command be
<strong>Path</strong>. We can also install other packages needed to run the tool.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/GoyalLab/singletCodeTools
<span class="n">Path</span> <span class="o">=</span> <span class="s2">&quot;path/to/singletCodeTools/repo&quot;</span>
<span class="o">%</span><span class="k">conda</span> install scipy tqdm matplotlib biopython python-levenshtein pandas
</pre></div>
</div>
</section>
<section id="next-step-is-to-understand-the-samples-present-in-the-fastq-files">
<h2>Next step is to understand the samples present in the FASTQ files.<a class="headerlink" href="#next-step-is-to-understand-the-samples-present-in-the-fastq-files" title="Permalink to this heading">¶</a></h2>
<p>The sample fastq files are in the inputFolder. We can identify the
sample name and number from the FASTQ file. For example,
sampleName_S1_L001_R1_001.fastq.gz means that the sample name is
sampleName and sample number is 1.Make sure that both read 1 and read 2
for each sample are present in the same folder (R1 and R2)</p>
</section>
<section id="creating-sample-sheet-for-these-two-samples">
<h2>Creating sample sheet for these two samples.<a class="headerlink" href="#creating-sample-sheet-for-these-two-samples" title="Permalink to this heading">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="s2">&quot;path/to/downloaded/and/unzipped/data&quot;</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sampleSheet</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/inputFiles/sampleSheet.csv&quot;</span><span class="p">)</span>
<span class="n">sampleSheet</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sampleName</th>
      <th>sampleNumber</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>sampleName</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>otherSampleName</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div><p>Now, to run the watermelon module of singletCodeTools, you need to run
this command. If we are going by the folder structure of the zipped file
and <strong>p</strong> is <em>path to the unzipped folder containing example files</em>,
then 1. <strong>inputFolder</strong> will be p/inputFiles/ 2. <strong>outputFolder</strong> will
be p/outputFiles/ 3. <strong>sampleSheet</strong> will be
p/inputFiles/sampleSheet.csv</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">subprocess</span>

<span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s1">&#39;python&#39;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">Path</span><span class="si">}</span><span class="s1">/commandLine/singletCodeCommandLine.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;watermelon&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-i&#39;</span><span class="p">,</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span>  <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/sampleSheet.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--outputName&#39;</span><span class="p">,</span> <span class="s1">&#39;watermelonBarcodeUmi.csv&#39;</span>
<span class="p">]</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Arguments</span> <span class="n">received</span><span class="p">:</span>
  <span class="n">command</span><span class="p">:</span> <span class="n">watermelon</span>
  <span class="n">inputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span>
  <span class="n">outputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">outputFiles</span>
  <span class="n">sampleSheet</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span><span class="n">sampleSheet</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">outputName</span><span class="p">:</span> <span class="n">watermelonBarcodeUmi</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">use10X</span><span class="p">:</span> <span class="kc">False</span>
  <span class="n">input10X</span><span class="p">:</span> <span class="kc">None</span>
<span class="n">All</span> <span class="n">the</span> <span class="n">inputs</span> <span class="k">for</span> <span class="n">the</span> <span class="n">command</span> <span class="n">are</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">will</span> <span class="n">proceed</span> <span class="k">with</span> <span class="n">creating</span> <span class="n">the</span> <span class="n">barcode</span> <span class="n">sheet</span> <span class="k">for</span> <span class="nb">all</span> <span class="n">the</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">sheet</span><span class="o">.</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">940</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">718</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
    <span class="s1">&#39;python&#39;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">Path</span><span class="si">}</span><span class="s1">/commandLine/singletCodeCommandLine.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;watermelon&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/sampleSheet.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--outputName&#39;</span><span class="p">,</span> <span class="s1">&#39;watermelonBarcodeUmiWith10X.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--use10X&#39;</span><span class="p">,</span>
    <span class="s1">&#39;--input10X&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/inputFiles/barcodes.tsv&#39;</span>
<span class="p">],</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Check if the command was successful</span>
<span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">returncode</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command executed successfully&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Output:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command failed&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Command</span> <span class="n">executed</span> <span class="n">successfully</span>
<span class="n">Output</span><span class="p">:</span>
 <span class="n">Arguments</span> <span class="n">received</span><span class="p">:</span>
  <span class="n">command</span><span class="p">:</span> <span class="n">watermelon</span>
  <span class="n">inputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span>
  <span class="n">outputFolder</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">outputFiles</span><span class="o">/</span>
  <span class="n">sampleSheet</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span><span class="n">sampleSheet</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">outputName</span><span class="p">:</span> <span class="n">watermelonBarcodeUmiWith10X</span><span class="o">.</span><span class="n">csv</span>
  <span class="n">use10X</span><span class="p">:</span> <span class="kc">True</span>
  <span class="n">input10X</span><span class="p">:</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">keerthana</span><span class="o">/</span><span class="n">Goyal_Lab</span><span class="o">/</span><span class="n">websiteToolData</span><span class="o">/</span><span class="n">thingsToAddToWebsite</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">watermelonVignetteData</span><span class="o">/</span><span class="n">inputFiles</span><span class="o">/</span><span class="n">barcodes</span><span class="o">.</span><span class="n">tsv</span>
<span class="n">All</span> <span class="n">the</span> <span class="n">inputs</span> <span class="k">for</span> <span class="n">the</span> <span class="n">command</span> <span class="n">are</span> <span class="n">valid</span> <span class="ow">and</span> <span class="n">will</span> <span class="n">proceed</span> <span class="k">with</span> <span class="n">creating</span> <span class="n">the</span> <span class="n">barcode</span> <span class="n">sheet</span> <span class="k">for</span> <span class="nb">all</span> <span class="n">the</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">sheet</span><span class="o">.</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">791</span>
<span class="n">Filtered</span> <span class="n">rows</span> <span class="n">of</span> <span class="n">dataframe</span><span class="p">:</span> <span class="mi">629</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">subprocess</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
    <span class="s1">&#39;python&#39;</span><span class="p">,</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">Path</span><span class="si">}</span><span class="s1">/commandLine/singletCodeCommandLine.py&#39;</span><span class="p">,</span>
    <span class="s1">&#39;count&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/watermelonBarcodeUmiWith10X.csv&#39;</span><span class="p">,</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/watermelon&#39;</span>
<span class="p">],</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Check if the command was successful</span>
<span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">returncode</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command executed successfully&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Output:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Command failed&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">result</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Command executed successfully
Output:
 Arguments received:
  command: count
  input_file: /home/keerthana/Goyal_Lab/websiteToolData/thingsToAddToWebsite/watermelonVignetteData/watermelonVignetteData/outputFiles/watermelonBarcodeUmiWith10X.csv
  out_prefix: /home/keerthana/Goyal_Lab/websiteToolData/thingsToAddToWebsite/watermelonVignetteData/watermelonVignetteData/outputFiles/watermelon
  umi_cutoff_ratio: 7.5e-06
  umi_diff_threshold: 50
  dominant_threshold: 10
  min_umi_good_data_cutoff: 2
INFO: Raw data counts
sampleNum
sampleName         693
otherSampleName    524
Name: count, dtype: int64
INFO: Using raio based filtering.
Current Sample Adjusted UMI cutoff: 2
Total cells: 45
Sample sampleName singlet: 43
Total Singlets: 43
Total Multiplets: 1
All singlets identified are unique? True
Total Singlets: 43
Total Multiplets: 1
INFO: Using raio based filtering.
Current Sample Adjusted UMI cutoff: 2
Total cells: 22
Sample otherSampleName singlet: 22
Total Singlets: 22
Total Multiplets: 0
All singlets identified are unique? True
Total Singlets: 22
Total Multiplets: 0
All singlets identified are unique? True
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_sampleName_singlets_stats.csv&quot;</span><span class="p">)</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#62575b&#39;</span><span class="p">,</span> <span class="s1">&#39;#2175a8&#39;</span><span class="p">,</span> <span class="s1">&#39;#feb422&#39;</span><span class="p">,</span> <span class="s1">&#39;#d62728&#39;</span><span class="p">,</span> <span class="s1">&#39;#d4d4d4&#39;</span><span class="p">]</span>  <span class="c1"># Example colors, modify as needed</span>
<span class="n">plotData</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="s1">&#39;total_cells&#39;</span><span class="p">,</span> <span class="s2">&quot;total_singlets&quot;</span><span class="p">])</span>

<span class="c1"># Plotting</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plotData</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;barh&#39;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>

<span class="k">for</span> <span class="n">plot</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">patches</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">plot</span><span class="o">.</span><span class="n">get_width</span><span class="p">())</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="p">(</span><span class="n">plot</span><span class="o">.</span><span class="n">get_x</span><span class="p">()</span> <span class="o">+</span> <span class="n">plot</span><span class="o">.</span><span class="n">get_width</span><span class="p">()</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">plot</span><span class="o">.</span><span class="n">get_y</span><span class="p">()</span> <span class="o">+</span> <span class="n">plot</span><span class="o">.</span><span class="n">get_height</span><span class="p">()</span><span class="o">/</span><span class="mi">2</span><span class="p">),</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Total cells&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Singlets by Criteria&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/watermelonDatasetVignette_10_0.png" src="../_images/watermelonDatasetVignette_10_0.png" />
<p>In the above plot, you see that the original data had 569 cells that
were removed due to low barcode UMI count, 43 singlets with a
single-barcode associated with them and 1 multiplet (singletCode could
not determine if it was a singlet for sure.)</p>
</section>
<section id="looking-at-the-scrnaseq-data-associated">
<h2>Looking at the scRNAseq data associated<a class="headerlink" href="#looking-at-the-scrnaseq-data-associated" title="Permalink to this heading">¶</a></h2>
<p>Since this data has both scRNAseq and barcodes for the same cells, we
can analyse them together</p>
<p>Installing and importing scanpy package to do this</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Install scanpy for further single-cell RNAseq analysis</span>
<span class="c1"># %conda install -c conda-forge scanpy python-igraph leidenalg</span>
<span class="c1">#Import scanpy</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
</pre></div>
</div>
<p>In case there are version conflicts during this installation or while
importing scanpy, we found <em>%conda update –all</em> to be an useful command
that fixed the version conflict previously. Reading in the 10X h5ad
object associated with the same watermelon data</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/inputFiles/watermelonScRnaSeqData.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1093 × 27264
</pre></div>
</div>
</section>
<section id="read-in-the-output-files-to-identify-cells-as-being-singlets-multiplets-or-being-removed-for-low-barcode-umi-threshold">
<h2>Read in the output files to identify cells as being singlets, multiplets or being removed for low barcode UMI threshold<a class="headerlink" href="#read-in-the-output-files-to-identify-cells-as-being-singlets-multiplets-or-being-removed-for-low-barcode-umi-threshold" title="Permalink to this heading">¶</a></h2>
<p>First, reading in the cellID-barcode-UMI sheet generated earlier with
additional filter using scRNAseq data</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cellidBarcodeUMI</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">/outputFiles/watermelonBarcodeUmiWith10X.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Reading in all the singlets and multiplets idenified in the two samples.
There might not always be multiplets - check the stats file to see if
there are any. In this example, there are no multiplets in
otherSampleName.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sampleNameSinglets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_sampleName_singlets_all.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">otherSampleNameSinglets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_otherSampleName_singlets_all.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">sampleNameMultiplets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelon_sampleName_multiplets.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<p>Identifying the cells that were below the barcode UMI threshold and were
filtered out by singletCode</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lowUmiCells</span> <span class="o">=</span> <span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">|</span>
                                 <span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">otherSampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">|</span>
                                 <span class="n">cellidBarcodeUMI</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameMultiplets</span><span class="p">[</span><span class="mi">0</span><span class="p">]))]</span>
</pre></div>
</div>
</section>
<section id="annotating-the-cells-in-adata-with-these-labels">
<h2>Annotating the cells in adata with these labels<a class="headerlink" href="#annotating-the-cells-in-adata-with-these-labels" title="Permalink to this heading">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Annotating the adata with these labels using the lists created</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;singlet&#39;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">otherSampleNameSinglets</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;singlet&#39;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sampleNameMultiplets</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;multiplet&#39;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">lowUmiCells</span><span class="p">[</span><span class="s1">&#39;cellID&#39;</span><span class="p">]),</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;low UMI&#39;</span>
</pre></div>
</div>
<p><strong>Note</strong> that in this vignette we are not doing any actual QC - but in
actual analysis, it would need to be done.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="calculating-pca-and-umap-for-visualization">
<h2>Calculating PCA and UMAP for visualization<a class="headerlink" href="#calculating-pca-and-umap-for-visualization" title="Permalink to this heading">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculating PCA for the data and plotting variance ratio</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/watermelonDatasetVignette_26_0.png" src="../_images/watermelonDatasetVignette_26_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;singletStatus&#39;</span><span class="p">],</span>
    <span class="n">size</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/watermelonDatasetVignette_27_0.png" src="../_images/watermelonDatasetVignette_27_0.png" />
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculating neighbours and UMAP from that for further visualization</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">random_state</span> <span class="o">=</span> <span class="mi">101010</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;singletStatus&#39;</span><span class="p">],</span>
    <span class="n">size</span><span class="o">=</span><span class="mi">60</span>
<span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/watermelonDatasetVignette_29_0.png" src="../_images/watermelonDatasetVignette_29_0.png" />
</section>
<section id="saving-the-final-adata">
<h2>Saving the final adata<a class="headerlink" href="#saving-the-final-adata" title="Permalink to this heading">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">/outputFiles/watermelonScRNA.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../infoCLI/" class="btn btn-neutral float-left" title="command line interface" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../singletCodePackageVignette/" class="btn btn-neutral float-right" title="Vignette to use singletCode package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Goyal Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>