<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>benchmarking doublet detection methods &mdash; singletCode 1.2 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/sphinx_rtd_size.css" type="text/css" />
      <link rel="stylesheet" href="../_static/colour.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script type="text/javascript">document.addEventListener('DOMContentLoaded', function() {document.querySelectorAll('a.reference.external').forEach(function(a) {a.setAttribute('target', '_blank'); a.setAttribute('rel', 'noopener noreferrer');});});</script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex/" />
    <link rel="search" title="Search" href="../search/" />
    <link rel="next" title="dataset information" href="../datasetsInfo/" />
    <link rel="prev" title="implementing singletCode on your data" href="../usageInformation/" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../" class="icon icon-home">
            singletCode
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search/" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../doubletInformation/">doublets in sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../singletCode/">what is singletCode?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../classifier/">classifier</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usageInformation/">implementing singletCode on your data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">benchmarking doublet detection methods</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#simulation-datasets">Simulation datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#benchmarking">Benchmarking</a></li>
<li class="toctree-l2"><a class="reference internal" href="#heterogeneity-effects">Heterogeneity effects</a></li>
<li class="toctree-l2"><a class="reference internal" href="#singletcode-for-scatac-seq-doublet-detection">singletCode for scATAC-seq doublet detection</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../datasetsInfo/">dataset information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../packageDocumentation/">singletCode PyPI package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../infoCLI/">command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../watermelonDatasetVignette/">Vignette to use singletCode Command Line tool to analyse single-cell data with watermelon barcodes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../singletCodePackageVignette/">Vignette to use singletCode package</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Contact/">questions? contact us!</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.cell.com/cell-genomics/fulltext/S2666-979X(24)00176-9">link to paper</a></li>
<li class="toctree-l1"><a class="reference external" href="https://www.goyallab.org/">Goyal Lab website</a></li>
<li class="toctree-l1"><a class="reference internal" href="../repositories/">repositories</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../">singletCode</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">benchmarking doublet detection methods</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/doubletDetectionMethods.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="benchmarking-doublet-detection-methods">
<h1>benchmarking doublet detection methods<a class="headerlink" href="#benchmarking-doublet-detection-methods" title="Permalink to this heading">¶</a></h1>
<section id="simulation-datasets">
<h2>Simulation datasets<a class="headerlink" href="#simulation-datasets" title="Permalink to this heading">¶</a></h2>
<p>To simulate doublets for benchmarking, we randomly selected the gene expression counts data from two cells that were found to be true singlets by singletCode. We averaged the counts from these two cells to generate simulated doublets and create datasets with various doublet percentages for benchmarking.</p>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="../_images/Figure1C.png"><img alt="Workflow schematic of how raw data is processed to generate true singlets based on a barcode UMI cutoff and doublet removal based on singletCode specifications, followed by the simulation of doublets for benchmarking doublet detection methods." src="../_images/Figure1C.png" style="width: 213.5px; height: 389.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Workflow schematic of how raw data is processed to generate true singlets based on a barcode UMI cutoff and doublet removal based on singletCode specifications, followed by the simulation of doublets for benchmarking doublet detection methods.</em></span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="benchmarking">
<h2>Benchmarking<a class="headerlink" href="#benchmarking" title="Permalink to this heading">¶</a></h2>
<dl class="simple">
<dt>We used these datasets to benchmark four doublet detection methods:</dt><dd><ol class="arabic simple">
<li><p><a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/scDblFinder.html">scDblFinder</a>,</p></li>
<li><p><a class="reference external" href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a>,</p></li>
<li><p><a class="reference external" href="https://github.com/AllonKleinLab/scrublet/">Scrublet</a> and</p></li>
<li><p><a class="reference external" href="https://github.com/kostkalab/scds">Hybrid</a></p></li>
</ol>
</dd>
</dl>
<p>We evaluated the AUPRC, AUROC, TNR, and doublet scores and calls of the four methods and found lower than expected performance for all methods. A plot of our results for AUPRC value is found below.</p>
<figure class="align-center" id="id2">
<a class="reference internal image-reference" href="../_images/Figure2A.png"><img alt="Color-coded boxplots of AUPRC value for all four doublet detection methods. Each boxplot is calculated from the AUPRC value after running the respective method on each sample of the dataset with actual doublet rate of 0.08 and expected doublet rate set to 0.05, 0.08, 0.1, 0.15, 0.2, and 0.25 where applicable. Dots represent the mean AUPRC value for a detection method grouped across all samples within a dataset. Lines represent standard deviation of the AUPRC values for a detection method across all samples within a dataset. Boxes span the first and third quartiles. Outliers are not depicted." src="../_images/Figure2A.png" style="width: 634.8000000000001px; height: 221.60000000000002px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Color-coded boxplots of AUPRC value for all four doublet detection methods. Each boxplot is calculated from the AUPRC value after running the respective method on each sample of the dataset with actual doublet rate of 0.08 and expected doublet rate set to 0.05, 0.08, 0.1, 0.15, 0.2, and 0.25 where applicable. Dots represent the mean AUPRC value for a detection method grouped across all samples within a dataset. Lines represent standard deviation of the AUPRC values for a detection method across all samples within a dataset. Boxes span the first and third quartiles. Outliers are not depicted.</em></span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>We examined the consistency of doublet labeling across different doublet detection methods by introducing a ‘similarity score’—a measure of the fraction of doublets identically classified by two methods. With an average similarity score of 0.66 across all datasets and methods, we observed variability in doublet detection.</p>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="../_images/Figure4A.png"><img alt="Pairwise similarity score between all benchmarked methods. A score of 1 indicates perfect identity and a 0 indicates complete disagreement. Average similarity score across all methods and datasets is 0.66." src="../_images/Figure4A.png" style="width: 670.5px; height: 492.0px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Pairwise similarity score between all benchmarked methods. A score of 1 indicates perfect identity and a 0 indicates complete disagreement. Average similarity score across all methods and datasets is 0.66.</em></span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>We further evaluated doublet detection on ensemble doublet detection methods (hybrid, Chord) and across sequencing technologies (10X Genomics, Smart-seq3). For a more detailed evaluation of our results, refer to our paper.</p>
</section>
<section id="heterogeneity-effects">
<h2>Heterogeneity effects<a class="headerlink" href="#heterogeneity-effects" title="Permalink to this heading">¶</a></h2>
<p>We wanted to know whether heterogeneity of a dataset affects the performance of doublet detection methods. Because heterogeneity can be impacted by many properties of a dataset, such as experimental design and data processing, we made conclusions based on heterogeneity within a sample. We did this by subsampling singlets and doublets within a single PC cluster for a sample (less heterogeneous, low Euclidean distance), and across all clusters for a sample (more heterogeneity, higher Euclidean distance).</p>
<figure class="align-default" id="id4">
<a class="reference internal image-reference" href="../_images/Figure3B.png"><img alt="Schematic of how transcriptionally similar and dissimilar cells are subsampled for subsequent within-sample heterogeneity determinations. A principal component analysis (left) was performed on each dataset and the first 30 PCs are used to identify Louvain clusters, represented by color. Transcriptionally similar (less heterogeneous) cells are identified from a single cluster and have a low Euclidean distance in PC space. Transcriptionally dissimilar (more heterogeneous) cells are identified from all clusters within the dataset and have a high average Euclidean distance in PC space." src="../_images/Figure3B.png" style="width: 766.5px; height: 171.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic of how transcriptionally similar and dissimilar cells are subsampled for subsequent within-sample heterogeneity determinations. A principal component analysis (left) was performed on each dataset and the first 30 PCs are used to identify Louvain clusters, represented by color. Transcriptionally similar (less heterogeneous) cells are identified from a single cluster and have a low Euclidean distance in PC space. Transcriptionally dissimilar (more heterogeneous) cells are identified from all clusters within the dataset and have a high average Euclidean distance in PC space.</em></span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
<section id="singletcode-for-scatac-seq-doublet-detection">
<h2>singletCode for scATAC-seq doublet detection<a class="headerlink" href="#singletcode-for-scatac-seq-doublet-detection" title="Permalink to this heading">¶</a></h2>
<p>We evaluated our method’s ability to assess doublet detection in scATAC-seq datasets using AMULET for a proof-of-concept analysis. We generated Watermelon-barcoded 10X Genomics Multiome datasets and applied AMULET to scATAC-seq fragments to categorize singlets and doublets. Concurrently, we identified true singlets in a barcoded scRNA-seq library from the same cells. By comparing the true negative and false positive rates between AMULET and singletCode across six datasets, we calculated AMULET’s average true negative rate (TNR) at 0.924. Our results demonstrate that singletCode can be used to benchmark doublet detection in other modalities besides scRNA-seq.</p>
<figure class="align-center" id="id5">
<a class="reference internal image-reference" href="../_images/Figure4G.png"><img alt="Schematic for 10X Single Cell Multiome dataset workflow resulting in TNR calculation. The same cells in the Multiome dataset undergo scATAC-seq, scRNA-seq, and barcode sequencing. Barcodes were processed with singletCode to identify ground truth singlets while the respective fragments file from ATAC was used to label doublets with AMULET. The AMULET-labeled singlets were compared with ground truth singletCode singlets to calculate TNR." src="../_images/Figure4G.png" style="width: 323.0px; height: 383.5px;" /></a>
<figcaption>
<p><span class="caption-text"><em>Schematic for 10X Single Cell Multiome dataset workflow resulting in TNR calculation. The same cells in the Multiome dataset undergo scATAC-seq, scRNA-seq, and barcode sequencing. Barcodes were processed with singletCode to identify ground truth singlets while the respective fragments file from ATAC was used to label doublets with AMULET. The AMULET-labeled singlets were compared with ground truth singletCode singlets to calculate TNR.</em></span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../usageInformation/" class="btn btn-neutral float-left" title="implementing singletCode on your data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../datasetsInfo/" class="btn btn-neutral float-right" title="dataset information" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Goyal Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>