===================
What is singletCode?
===================

singletCode is a framework to extract true singlets from barcoded scRNA-seq data. Our pipeline identified barcoded singlets from 10 different publications and 6 original experimental scRNA-seq samples generated in this study, encompassing 7 different barcoding technologies, 149 scRNA-seq samples, 3 unique sequencing technologies, and a total of 564,579 cells. Of the 338,948 barcoded cells, we extracted 293,618.(read more about the barcoded datasets we used singletCode with :ref:`here<datasetInfo>`). 


Singlet identification
-----------------------------------
singletCode leverages the fact that the lineage barcodes are present within the cell before sequencing. As a result of multiplicity of infection, singletCode identifies cells which meet any of the following conditions as a true singlet:

   #. 1 barcode/1 cell ID
   #. >1 barcode per 1 cell ID, but 1 barcode has significantly more UMI counts than the other barcodes within the same cell
   #. M barcodes per 1 cell ID, but the same combination of M barcodes are found in other cells in the same sample
   #. M barcodes per 1 cell ID, but the same combination of M barcodes are found in other cells across samples within the same experimental design (common in barcoding studies) 


.. figure:: /images/Figure1F.png
   :scale: 50 %
   :alt: Stacked bar chart showing types of singlets in each dataset. Each color represents one type of barcode composition that constitutes a singlet. Single barcode means one cell ID has one associated lineage barcode. Dominant barcode means one barcode has the highest UMI counts in one cell. Multiple barcodes within or across samples are scenarios where the same barcode combinations are observed in the same sample or across twin samples. See our paper for details.

   
   *Stacked bar chart showing types of singlets in each dataset. Each color represents one type of barcode composition that constitutes a singlet. Single barcode means one cell ID has one associated lineage barcode. Dominant barcode means one barcode has the highest UMI counts in one cell. Multiple barcodes within or across samples are scenarios where the same barcode combinations are observed in the same sample or across twin samples. See our paper for details.*

.. _simulationInfo:

Simulation Datasets 
---------------------------------------
To simulate the doublets, we randomly selected the gene expression counts data from two cells we identify as true singlets. We averaged or summed the counts from these two cells to generate simulated doublets.  

For a desired doublet percentage, we generated an initial number of doublets based on the initial assumption that doublets are generated from singlets without replacement such that, for each doublet generated, the total number of cells will decrease by 1. Then, it follows that:  `number of doublets = (doublet percentage)*(number of initial singlets - number of doublets)`.

We generated the number of doublets determined by this formula. Next, we removed the singlet cells used to create doublets and corrected for generating doublets from singlets with replacement to ensure the desired doublet percentage.
  
The final number of singlets is given by  `number of singlets = number of doublets*((1 - doublet percentage)/(doublet percentage))` and we subsampled that number of singlets to be included in the final matrix. This approach minimized the number of cells that needed to be trimmed in the doublet generation process to maintain desired doublet percentages. 
In the rare cases where there were not enough singlets remaining after those used to create doublets were removed, the number of doublets was instead calculated according to  `number of doublets = doublet percentage * (number of singlets/(1 - doublet percentage))`.

.. figure:: /images/singletCode-2.png
   :scale: 50 %
   :alt: Method to simulate doublets using true singlets from singletCode
   
   Method to simulate doublets using true singlets from singletCode

The final scRNA-seq datasets are generated by adding such simulated doublets into the datasets at different percentages (10-40%).
 
.. contents:: Contents:
   :local: